---
title: "Tidyverse: TidyR and ggplot2"
author: "Gamaliel Cabria"
date: "`r Sys.Date()`"
output: html_document
---

# Tidying Data with `tidyr`
The `tidyr` package is designed to help you create tidy data, which is a standardized way of organizing data that makes it easier to work with. Tidy data has the following characteristics:
- Each variable forms a column.
- Each observation forms a row.
- Each type of observational unit forms a table.
  
Tidy data is essential for efficient data analysis and visualization, as it allows you to easily manipulate and visualize your data using the tidyverse packages.

## Example Dataset with `tidyr`
In this section, we will use another Bioinformatics dataset to demonstrate the use of `tidyr` for tidying data. The dataset is a summarised table listing the instances that the program `virsorter` predicted a viral sequence in a metagenomic dataset. The dataset contains the following columns:
- **Genome**: The name of the genome file.
  - It is divided by the **Site**, **Timepoint** (i.e. SE046), **Sample Name** (i.e. G07_cluster_2), and **Population**(i.e. Blue).
- **Viral Type**: The predicted viral sequence is either **dsDNAPhage** or **ssDNAPhage**.

To download the data, run the following code:
```bash
wget https://raw.githubusercontent.com/gamalielcabria/gamalielcabria.github.io/main/DataAnalysis/files/virsorter/virsorter_output.csv
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
setwd("/home/gam/github/gamalielcabria.github.io/DataAnalysis")
```

Let us load the dataset and take a look at its structure:
```{r echo=TRUE, warning=FALSE, message=FALSE}
library(tidyverse)

virsorter_data <- read_csv("./virsorter_output.csv",
  col_types = cols(
    ...1 = col_skip(),              # Skip the first unnamed column
    Genome = col_character(),       # The genome file name
    Phage_Type = col_character()    # The predicted viral sequence type
  )
)
head(virsorter_data)
```

---

## Separate: Expanding a column into multiple columns
Sometimes, a single column contains multiple pieces of information that can be separated into distinct columns. The `separate()` function from `tidyr` allows you to do this easily.

In our dataset, the **Genome** column contains multiple pieces of information separated by underscores. We can use `separate()` to split this column into multiple columns.
```{r echo=TRUE, warning=FALSE, message=FALSE}
virsorter_data_separated <- virsorter_data %>%
  separate(Genome, 
            into = c("Site", "Timepoint", "Sample_Name"), 
            sep = "_", 
            remove = TRUE,
            extra = "merge",
            fill = "right"
            )
head(virsorter_data_separated, n = 20)
```

You might notice that the **Population** column is still missing. It is hard to extract since some genomes (i.e. KRP1_SE055_G12_cluster_2_cluster_1_Green_vs2) have more than one extra sets of columns due to extra '_'. 

We can use the `str_extract()` function from the `stringr` package to extract the population from the **Sample_Name** column.

The `extra = "merge"` argument in the `separate()` function allows us to merge any extra pieces of information into the last column, which is useful when the number of pieces of information varies across rows. The `fill = "right"` argument ensures that any missing values are filled with `NA` on the right side.
<br>

### The opposite of `separate()`: `unite()`
The `unite()` function from `tidyr` allows you to combine multiple columns into a single column. This is useful when you want to create a new column that contains information from multiple existing columns.
```{r echo=TRUE, warning=FALSE, message=FALSE}
virsorter_data_united <- virsorter_data_separated %>%
  unite("Genome", Site, Timepoint, Sample_Name, sep = "_", remove = FALSE)
head(virsorter_data_united)
```

### Featuring: Tidyverse's `stringr` 

The `stringr` package provides a set of functions for working with strings in R. It is part of the tidyverse and is designed to make string manipulation easier and more consistent.
```{r echo=TRUE, warning=FALSE, message=FALSE}
library(stringr)
virsorter_data_separated <- virsorter_data_united %>%
  mutate(Population = str_extract(Genome, "Green|Blue|Yellow|Orange|Brown|Purple"))
head(virsorter_data_separated, n = 20)
```

{:.notice}
> The `mutate()` function is still needed to create a new column in the dataset. This is because `str_extract()` returns a character vector of the same length as the input, and we need to assign it to a new column in the dataset.

---

## Recode: Replacing values in a column
Sometimes, you may want to replace specific values in a column with new values. The `recode()` function from `dplyr` allows you to do this easily.
```{r echo=TRUE, warning=FALSE, message=FALSE}
virsorter_data_recode <- virsorter_data_separated %>% 
    mutate(Population = recode(Population, 
                    Orange="Hydrogenobacter sp. Cluster 2 Orange", 
                    Blue="Hydrogenobacter sp. Cluster 2 Blue", 
                    Green="Hydrogenobacter sp. Cluster 1 Green", 
                    Yellow="H. thermophilus", 
                    Brown="Sulfurihydrogenibium sp.",
                    Purple="Venenivibrio sp."
                    )
    ) 
head(virsorter_data_recode)
```

Let us create a summary table that counts the number of viral sequences per Phage_Type and Genome:
```{r echo=TRUE, warning=FALSE, message=FALSE}
virsorter_summary <- virsorter_data_recode %>%
  group_by(Population, Phage_Type) %>%
  summarise(Count = n()
  ) %>%
  arrange(Population, Phage_Type)
virsorter_summary
```

---

## Converting long data to wide format: `pivot_wider`
The table `virsorter_summary` is in long format, where each row represents a unique combination of Population and Phage_Type. We can use the `pivot_wider()` function from `tidyr` to convert this table into wide format, where each Population has its own column for each Phage_Type.
```{r echo=TRUE, warning=FALSE, message=FALSE}      
virsorter_wide <- virsorter_summary %>%
  pivot_wider(names_from = Phage_Type, values_from = Count, values_fill = 0)
head(virsorter_wide)
```

Wider data formats are often easier to read and can be useful for certain types of analyses, such as when you want to compare values across different categories.

### Converting wide data to long format: `pivot_longer`
Meanwhile, if you have a dataset in wide format and you want to convert it back to long format, you can use the `pivot_longer()` function. This is useful when you want to perform operations that require tidy data.

```{r echo=TRUE, warning=FALSE, message=FALSE}
virsorter_long <- virsorter_wide %>%
  pivot_longer(cols = -Population, 
               names_to = "Phage_Type", 
               values_to = "Count")
head(virsorter_long)
```

{:.warning}
> The `pivot_longer()` function is the opposite of `pivot_wider()`. It might seem reversible, but it is not always straightforward. The `cols` argument specifies which columns to pivot, and the `names_to` and `values_to` arguments specify the names of the new columns created from the pivoting operation.
>
> When there is missing data, the `pivot_longer()` function will fill in the missing values with `NA`. This is important to keep in mind when working with datasets that may have missing values. 

{:.info}
> **`replace_na()`**:
> It can be manually replaced with `0` or any other value using the `replace_na()` function from `tidyr`.
> >`df_long <- df_long %>% mutate(score = replace_na(score, 0))`

---

## Expanding a dataset with `expand()`
Let us subset the dataset from `virsorter_data_recode` to only include the **Population** and **Site** columns. We can then use the `expand()` function to create a dataset that contains all combinations of these two variables.
```{r echo=TRUE, warning=FALSE, message=FALSE}
virsorter_expanded <- virsorter_data_recode %>%
    select(Population, Site) %>%
    distinct() %>%
    expand(Population, Site)
virsorter_expanded
```

Let us compare the original dataset with the expanded dataset. The original dataset has 18 rows, while the expanded dataset has 30 rows. This is because the expanded dataset contains all combinations of Population, Site, and Phage_Type, including those that were not present in the original dataset.
```{r echo=TRUE, warning=FALSE, message=FALSE}
virsorter_original <- virsorter_data_recode %>%
    select(Population, Site) %>%
    distinct()
virsorter_original
```

As you can see, the expanded dataset contains additional rows that were not present in the original dataset. This is useful when you want to ensure that all combinations of variables are represented in your data, which can be important for certain types of analyses or visualizations.

{:.note}
> The `expand()` function is particularly useful when you want to create a dataset that contains all possible combinations of variables, even if some combinations do not exist in the original dataset. This can be helpful for creating visualizations or performing analyses that require complete datasets.

---

## Complete a dataset with `complete()`
The `complete()` function from `tidyr` allows you to fill in missing combinations of variables in your dataset. This is useful when you want to ensure that all possible combinations of variables are represented in your data.
```{r echo=TRUE, warning=FALSE, message=FALSE}
virsorter_incomplete <- virsorter_data_recode %>%
    group_by(Population, Phage_Type, Site, Timepoint) %>%
    summarise(Count = n()) %>%
    ungroup() %>%
    pivot_wider(names_from = Phage_Type, values_from = Count, values_fill = 0)
virsorter_incomplete
```

The result shows 18 lines. However, we can see that not all sites and Timepoint are present in the dataset. For example, there are no entries for **SE046** in the **Green** population. We can use the `complete()` function to fill in these missing combinations with `0` counts.

```{r echo=TRUE, warning=FALSE, message=FALSE}
virsorter_complete <- virsorter_data_recode %>%
    group_by(Population, Phage_Type, Site, Timepoint) %>%
    summarise(Count = n()) %>%
    ungroup() %>%
    complete(Population, Phage_Type, Site, Timepoint, fill = list(Count = 0)) %>%
    pivot_wider(names_from = Phage_Type, values_from = Count, values_fill = 0)
head(virsorter_complete, n = 20)
```

The complete dataset now has 90 rows, with all combinations of Population, Phage_Type, Site, and Timepoint represented. The missing combinations have been filled with `0` counts.

The `complete()` function is particularly useful when you want to ensure that your dataset is fully populated with all possible combinations of variables, which can be important for certain types of analyses or visualizations.

---

# Visualizing Data with `ggplot2`
The `ggplot2` package is a powerful and flexible system for creating static graphics in R. It is part of the tidyverse and is designed to work seamlessly with tidy data.

## Example Visualization: Bar Plot of Viral Types by Population
We can create a bar plot to visualize the distribution of viral types by population using `ggplot2`. The `geom_bar()` function is used to create bar plots, and we can use the `fill` aesthetic to color the bars by viral type.
```{r echo=TRUE, warning=FALSE, message=FALSE}  
virsorter_input <- virsorter_summary %>%
    ungroup() %>%
    complete(Population, Phage_Type, fill = list(Count = 0))

plot <- ggplot(virsorter_input) +
  geom_bar(aes( x = Population, 
                y = Count, 
                fill = Phage_Type
                ),
                stat = "identity",
                position = "dodge"
        ) +
  labs(title = "Distribution of Viral Types by Population",
       x = "Population",
       y = "Count") +
  theme_minimal()
plot
```

## Plot Geom Objects
Let us make a time series from our virsorter dataset.

First, let us re-analyse it so that viral count is show by timepoint and population.
```{r echo=TRUE, warning=FALSE, message=FALSE}
virsorter_time_series <- virsorter_data_recode %>%
  group_by(Population, Timepoint, Phage_Type) %>%
  summarise(Count = n(), .groups = 'drop') %>%
  arrange(Population, Timepoint, Phage_Type) %>%
  ungroup() %>%
  mutate(Timepoint = recode(Timepoint, 
         SE046 = 1,
         SE053 = 2,
         SE055 = 3,
         SE056 = 4,
         SE057 = 5,
         SE058 = 6,
         )
  ) %>%
  complete(Population, Timepoint, Phage_Type, fill = list(Count = 0)) 
virsorter_time_series
```

Let us plot the time series data using `ggplot2`. We will use `geom_line()` to create a line plot and `geom_point()` to add points for each timepoint.
```{r echo=TRUE, warning=FALSE, message=FALSE}
virsorter_time_series_orange <- virsorter_time_series %>%
  filter(Population == "Hydrogenobacter sp. Cluster 2 Orange")

plot <- ggplot(virsorter_time_series_orange, aes(x = Timepoint, y = Count, color = Phage_Type)) +
  geom_line(aes(group = Phage_Type), linewidth = 1) +
  geom_point(size = 3) +
  labs(title = "Viral Types Over Time by Population",
       x = "Timepoint",
       y = "Count") +
  scale_x_continuous(breaks = 1:6, labels = c("SE046", "SE053", "SE055", "SE056", "SE057", "SE058")) +
  theme_minimal() +
  theme(legend.position = "bottom")
plot
```

We filter the dataset to only include the **Hydrogenobacter sp. Cluster 2 Orange** population. The `geom_line()` function is used to create lines connecting the points for each viral type, and `geom_point()` adds points at each timepoint.

## Plotting with `facet_wrap()`
We can also use `facet_wrap()` to create separate plots for each population. This allows us to visualize the distribution of viral types within each population.
```{r echo=TRUE, warning=FALSE, message=FALSE}
plot <- ggplot(virsorter_time_series, aes(x = Timepoint, y = Count, color = Phage_Type)) +
  geom_line(aes(group = Phage_Type), linewidth = 1) +
  geom_point(size = 3) +
  labs(title = "Viral Types Over Time by Population",
       x = "Timepoint",
       y = "Count") +
  scale_x_continuous(breaks = 1:6, labels = c("SE046", "SE053", "SE055", "SE056", "SE057", "SE058")) +
  theme_minimal() +
  theme(legend.position = "bottom") +
  facet_wrap(~ Population)
plot
```

If you want to see the non-facet_wrap version, you can use the `facet_null()` function to remove the facets.
```{r echo=TRUE, warning=FALSE, message=FALSE}  
plot <- ggplot(virsorter_time_series, aes(x = Timepoint, y = Count, color = Phage_Type)) +
  geom_line(aes(group = Phage_Type), linewidth = 1) +
  geom_point(size = 3) +
  labs(title = "Viral Types Over Time by Population",
       x = "Timepoint",
       y = "Count") +
  scale_x_continuous(breaks = 1:6, labels = c("SE046", "SE053", "SE055", "SE056", "SE057", "SE058")) +
  theme_minimal() +
  theme(legend.position = "bottom") 
plot
```


## Changing the Color Palette
We can customize the color palette of our plot using the `scale_color_manual()` function. This allows us to specify the colors for each viral type.
```{r echo=TRUE, warning=FALSE, message=FALSE}
plot <- ggplot(virsorter_time_series, aes(x = Timepoint, y = Count, color = Phage_Type)) +
  geom_line(aes(group = Phage_Type), linewidth = 1) +
  geom_point(size = 3) +
  labs(title = "Viral Types Over Time by Population",
       x = "Timepoint",
       y = "Count") +
  scale_x_continuous(breaks = 1:6, labels = c("SE046", "SE053", "SE055", "SE056", "SE057", "SE058")) +
  scale_color_manual(values = c("dsDNAphage" = "#FF5733", "ssDNA" = "#33FF57")) +
  theme_minimal() +
  theme(legend.position = "bottom") +
  facet_wrap(~ Population)
plot
```

Let us try the Wes Anderson color palette from the `wesanderson` package. This package provides a set of aesthetically pleasing color palettes inspired by the films of Wes Anderson.
```{r echo=TRUE, warning=FALSE, message=FALSE}
library(wesanderson) # Install the package if you haven't already

plot <- ggplot(virsorter_time_series, aes(x = Timepoint, y = Count, color = Phage_Type)) +
  geom_line(aes(group = Phage_Type), linewidth = 1) +
  geom_point(size = 3) +
  labs(title = "Viral Types Over Time by Population",
       x = "Timepoint",
       y = "Count") +
  scale_x_continuous(breaks = 1:6, labels = c("SE046", "SE053", "SE055", "SE056", "SE057", "SE058")) +
  scale_color_manual(values = wes_palette("GrandBudapest1",2)) + # Use a Wes Anderson color palette
  theme_minimal() +
  theme(legend.position = "bottom") +
  facet_wrap(~ Population)
plot
```

---

## The Themes of `ggplot2`
You can customize the appearance of your plots using themes in `ggplot2`. The `theme_minimal()` function is one of the built-in themes that provides a clean and minimalistic look. You can also create your own custom themes or use other built-in themes like `theme_bw()`, `theme_classic()`, etc.
```{r echo=TRUE, warning=FALSE, message=FALSE}
plot <- ggplot(virsorter_time_series, aes(x = Timepoint, y = Count, color = Phage_Type)) +
  geom_line(aes(group = Phage_Type), linewidth = 1) +
  geom_point(size = 3) +
  labs(title = "Viral Types Over Time by Population",
       x = "Timepoint",
       y = "Count") +
  scale_x_continuous(breaks = 1:6, labels = c("SE046", "SE053", "SE055", "SE056", "SE057", "SE058")) +
  scale_color_manual(values = wes_palette("GrandBudapest1",2)) + # Use a Wes Anderson color palette
  theme_minimal() + # Use a minimal theme
  theme(legend.position = "bottom",
        axis.title.x = element_text(size = 16, face = "bold"),
        axis.title.y = element_text(size = 15, face = "bold"),
        strip.text.x = element_text(size = 15, face = "bold"),
        plot.title = element_text(size = 18, face = "bold", hjust = 0.5)
        ) +
  facet_wrap(~ Population)
plot
```

Other themes can be applied to the plot using the `theme()` function. You can customize various aspects of the plot, such as axis labels, titles, and legend positions.

Other important theme arguments include:
## Important `theme()` Arguments in ggplot2

| Argument                | What it Controls                                 | Example Usage                                 |
|-------------------------|--------------------------------------------------|-----------------------------------------------|
| `legend.position`       | Position of the legend (`"right"`, `"bottom"`, `"left"`, `"top"`, or `"none"`) | `theme(legend.position = "bottom")`           |
| `axis.title.x`          | X-axis title appearance                          | `theme(axis.title.x = element_text(size = 14))`|
| `axis.title.y`          | Y-axis title appearance                          | `theme(axis.title.y = element_text(size = 14))`|
| `axis.text.x`           | X-axis tick labels                               | `theme(axis.text.x = element_text(angle = 45))`|
| `axis.text.y`           | Y-axis tick labels                               | `theme(axis.text.y = element_text(size = 12))` |
| `plot.title`            | Plot title appearance                            | `theme(plot.title = element_text(hjust = 0.5))`|
| `strip.text.x`          | Facet label appearance (x)                       | `theme(strip.text.x = element_text(size = 12))`|
| `strip.text.y`          | Facet label appearance (y)                       | `theme(strip.text.y = element_text(face = "bold"))`|
| `panel.grid.major`      | Major grid lines                                 | `theme(panel.grid.major = element_blank())`    |
| `panel.grid.minor`      | Minor grid lines                                 | `theme(panel.grid.minor = element_blank())`    |
| `panel.background`      | Plot panel background                            | `theme(panel.background = element_rect(fill = "white"))`|
| `plot.background`       | Entire plot background                           | `theme(plot.background = element_rect(fill = "gray90"))`|
| `legend.title`          | Legend title appearance                          | `theme(legend.title = element_text(face = "bold"))`|
| `legend.text`           | Legend text appearance                           | `theme(legend.text = element_text(size = 10))` |

These arguments allow you to control the appearance of almost every aspect of your ggplot2 visualizations.


# Summary Table: Key Data Tidying Functions

| Function         | Purpose                                                      | Example Usage                                                        |
|------------------|-------------------------------------------------------------|----------------------------------------------------------------------|
| `read_csv()`     | Read CSV files, skip columns                                | `read_csv("file.csv", col_types = cols(...1 = col_skip()))`          |
| `separate()`     | Split one column into multiple columns                      | `separate(df, col, into = c("A", "B"), sep = "_")`                   |
| `unite()`        | Combine multiple columns into one                           | `unite(df, "new_col", A, B, sep = "_")`                              |
| `mutate()`       | Create or modify columns                                    | `mutate(df, new_col = old_col * 2)`                                  |
| `str_extract()`  | Extract patterns from strings                               | `mutate(df, col2 = str_extract(col1, "pattern"))`                    |
| `recode()`       | Replace specific values in a column                         | `mutate(df, col = recode(col, old = "new"))`                         |
| `group_by()`     | Group data for summary operations                           | `group_by(df, col)`                                                  |
| `summarise()`    | Summarize data (e.g., count, mean)                          | `summarise(df, n = n())`                                             |
| `pivot_wider()`  | Convert long data to wide format                            | `pivot_wider(df, names_from, values_from)`                           |
| `pivot_longer()` | Convert wide data to long format                            | `pivot_longer(df, cols, names_to, values_to)`                        |
| `replace_na()`   | Replace missing values                                      | `mutate(df, col = replace_na(col, 0))`                               |
| `expand()`       | Create all combinations of variables                        | `expand(df, col1, col2)`                                             |
| `complete()`     | Fill in missing combinations of variables                   | `complete(df, col1, col2, fill = list(count = 0))`                   |