<!DOCTYPE html> <html lang="en-US"> <head> <meta charset="UTF-8"> <meta http-equiv="X-UA-Compatible" content="IE=Edge"> <link rel="stylesheet" href="/assets/css/just-the-docs-default.css"> <link rel="stylesheet" href="/assets/css/just-the-docs-head-nav.css" id="jtd-head-nav-stylesheet"> <style id="jtd-nav-activation"> .site-nav > ul.nav-list:first-child > li > a, .site-nav > ul.nav-list:first-child > li > ul > li:not(:nth-child(2)) > a, .site-nav > ul.nav-list:first-child > li > ul > li > ul > li a { background-image: none; } .site-nav > ul.nav-list:not(:first-child) a, .site-nav li.external a { background-image: none; } .site-nav > ul.nav-list:first-child > li:nth-child(3) > ul > li:nth-child(2) > a { font-weight: 600; text-decoration: none; }.site-nav > ul.nav-list:first-child > li:nth-child(3) > button svg, .site-nav > ul.nav-list:first-child > li:nth-child(3) > ul > li:nth-child(2) > button svg { transform: rotate(-90deg); }.site-nav > ul.nav-list:first-child > li.nav-list-item:nth-child(3) > ul.nav-list, .site-nav > ul.nav-list:first-child > li.nav-list-item:nth-child(3) > ul.nav-list > li.nav-list-item:nth-child(2) > ul.nav-list { display: block; } </style> <script src="/assets/js/vendor/lunr.min.js"></script> <script src="/assets/js/just-the-docs.js"></script> <meta name="viewport" content="width=device-width, initial-scale=1"> <!-- Begin Jekyll SEO tag v2.8.0 --> <title>Lesson 2 - Basic Operations in Terminal and BASH | üß¨ MEB Training Suite</title> <meta name="generator" content="Jekyll v4.3.4" /> <meta property="og:title" content="Lesson 2 - Basic Operations in Terminal and BASH" /> <meta property="og:locale" content="en_US" /> <meta name="description" content="Bookmark this to keep an eye on my project updates!" /> <meta property="og:description" content="Bookmark this to keep an eye on my project updates!" /> <link rel="canonical" href="http://localhost:4000/Intro2Bioinfo/Lesson2.html" /> <meta property="og:url" content="http://localhost:4000/Intro2Bioinfo/Lesson2.html" /> <meta property="og:site_name" content="üß¨ MEB Training Suite" /> <meta property="og:type" content="website" /> <meta name="twitter:card" content="summary" /> <meta property="twitter:title" content="Lesson 2 - Basic Operations in Terminal and BASH" /> <script type="application/ld+json"> {"@context":"https://schema.org","@type":"WebPage","description":"Bookmark this to keep an eye on my project updates!","headline":"Lesson 2 - Basic Operations in Terminal and BASH","url":"http://localhost:4000/Intro2Bioinfo/Lesson2.html"}</script> <!-- End Jekyll SEO tag --> </head> <body> <a class="skip-to-main" href="#main-content">Skip to main content</a> <svg xmlns="http://www.w3.org/2000/svg" class="d-none"> <symbol id="svg-link" viewBox="0 0 24 24"> <title>Link</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-link"> <path d="M10 13a5 5 0 0 0 7.54.54l3-3a5 5 0 0 0-7.07-7.07l-1.72 1.71"></path><path d="M14 11a5 5 0 0 0-7.54-.54l-3 3a5 5 0 0 0 7.07 7.07l1.71-1.71"></path> </svg> </symbol> <symbol id="svg-menu" viewBox="0 0 24 24"> <title>Menu</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-menu"> <line x1="3" y1="12" x2="21" y2="12"></line><line x1="3" y1="6" x2="21" y2="6"></line><line x1="3" y1="18" x2="21" y2="18"></line> </svg> </symbol> <symbol id="svg-arrow-right" viewBox="0 0 24 24"> <title>Expand</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-chevron-right"> <polyline points="9 18 15 12 9 6"></polyline> </svg> </symbol> <!-- Feather. MIT License: https://github.com/feathericons/feather/blob/master/LICENSE --> <symbol id="svg-external-link" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-external-link"> <title id="svg-external-link-title">(external link)</title> <path d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"></path><polyline points="15 3 21 3 21 9"></polyline><line x1="10" y1="14" x2="21" y2="3"></line> </symbol> <symbol id="svg-doc" viewBox="0 0 24 24"> <title>Document</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-file"> <path d="M13 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V9z"></path><polyline points="13 2 13 9 20 9"></polyline> </svg> </symbol> <symbol id="svg-search" viewBox="0 0 24 24"> <title>Search</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-search"> <circle cx="11" cy="11" r="8"></circle><line x1="21" y1="21" x2="16.65" y2="16.65"></line> </svg> </symbol> <!-- Bootstrap Icons. MIT License: https://github.com/twbs/icons/blob/main/LICENSE.md --> <symbol id="svg-copy" viewBox="0 0 16 16"> <title>Copy</title> <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-clipboard" viewBox="0 0 16 16"> <path d="M4 1.5H3a2 2 0 0 0-2 2V14a2 2 0 0 0 2 2h10a2 2 0 0 0 2-2V3.5a2 2 0 0 0-2-2h-1v1h1a1 1 0 0 1 1 1V14a1 1 0 0 1-1 1H3a1 1 0 0 1-1-1V3.5a1 1 0 0 1 1-1h1v-1z"/> <path d="M9.5 1a.5.5 0 0 1 .5.5v1a.5.5 0 0 1-.5.5h-3a.5.5 0 0 1-.5-.5v-1a.5.5 0 0 1 .5-.5h3zm-3-1A1.5 1.5 0 0 0 5 1.5v1A1.5 1.5 0 0 0 6.5 4h3A1.5 1.5 0 0 0 11 2.5v-1A1.5 1.5 0 0 0 9.5 0h-3z"/> </svg> </symbol> <symbol id="svg-copied" viewBox="0 0 16 16"> <title>Copied</title> <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-clipboard-check-fill" viewBox="0 0 16 16"> <path d="M6.5 0A1.5 1.5 0 0 0 5 1.5v1A1.5 1.5 0 0 0 6.5 4h3A1.5 1.5 0 0 0 11 2.5v-1A1.5 1.5 0 0 0 9.5 0h-3Zm3 1a.5.5 0 0 1 .5.5v1a.5.5 0 0 1-.5.5h-3a.5.5 0 0 1-.5-.5v-1a.5.5 0 0 1 .5-.5h3Z"/> <path d="M4 1.5H3a2 2 0 0 0-2 2V14a2 2 0 0 0 2 2h10a2 2 0 0 0 2-2V3.5a2 2 0 0 0-2-2h-1v1A2.5 2.5 0 0 1 9.5 5h-3A2.5 2.5 0 0 1 4 2.5v-1Zm6.854 7.354-3 3a.5.5 0 0 1-.708 0l-1.5-1.5a.5.5 0 0 1 .708-.708L7.5 10.793l2.646-2.647a.5.5 0 0 1 .708.708Z"/> </svg> </symbol> </svg> <div class="side-bar"> <div class="site-header" role="banner"> <a href="/" class="site-title lh-tight"> üß¨ MEB Training Suite </a> <button id="menu-button" class="site-button btn-reset" aria-label="Toggle menu" aria-pressed="false"> <svg viewBox="0 0 24 24" class="icon" aria-hidden="true"><use xlink:href="#svg-menu"></use></svg> </button> </div> <nav aria-label="Main" id="site-nav" class="site-nav"> <ul class="nav-list"><li class="nav-list-item"><a href="/" class="nav-list-link">Welcome</a></li><li class="nav-list-item"><a href="/getting-started.html" class="nav-list-link">Getting Started</a></li><li class="nav-list-item"><button class="nav-list-expander btn-reset" aria-label="toggle items in Intro to Bioinfo category" aria-pressed="false"> <svg viewBox="0 0 24 24" aria-hidden="true"><use xlink:href="#svg-arrow-right"></use></svg> </button><a href="/Intro2Bioinfo/" class="nav-list-link">Intro to Bioinfo</a><ul class="nav-list"><li class="nav-list-item"><a href="/Intro2Bioinfo/Lesson1.html" class="nav-list-link">Lesson 1 - The Terminal, the OS, and Bioinfo</a></li><li class="nav-list-item"><a href="/Intro2Bioinfo/Lesson2.html" class="nav-list-link">Lesson 2 - Basic Operations in Terminal and BASH</a></li><li class="nav-list-item"><a href="/Intro2Bioinfo/Lesson3.html" class="nav-list-link">Lesson 3 - The commands to rule them all</a></li><li class="nav-list-item"><a href="/Intro2Bioinfo/Exercise3.html" class="nav-list-link">Exercise 3 - Useful Commands</a></li><li class="nav-list-item"><a href="/Intro2Bioinfo/Lesson4.html" class="nav-list-link">Lesson 4 - Installing New Commands</a></li><li class="nav-list-item"><a href="/Intro2Bioinfo/Lesson5.html" class="nav-list-link">Lesson 5 - BASH Scripting</a></li><li class="nav-list-item"><a href="/Intro2Bioinfo/Lesson6.html" class="nav-list-link">Lesson 6 - Intro to ARC and SLURM JOB submissions</a></li></ul></li></ul> </nav> <footer class="site-footer"> This site uses <a href="https://github.com/just-the-docs/just-the-docs">Just the Docs</a>, a documentation theme for Jekyll. </footer> </div> <div class="main" id="top"> <div id="main-header" class="main-header"> <div class="search" role="search"> <div class="search-input-wrap"> <input type="text" id="search-input" class="search-input" tabindex="0" placeholder="Search üß¨ MEB Training Suite" aria-label="Search üß¨ MEB Training Suite" autocomplete="off"> <label for="search-input" class="search-label"><svg viewBox="0 0 24 24" class="search-icon"><use xlink:href="#svg-search"></use></svg></label> </div> <div id="search-results" class="search-results"></div> </div> <nav aria-label="Auxiliary" class="aux-nav"> <ul class="aux-nav-list"> <li class="aux-nav-list-item"> <a href="https://github.com/gamalielcabria" class="site-button" > Template Repository </a> </li> </ul> </nav> </div> <div class="main-content-wrap"> <nav aria-label="Breadcrumb" class="breadcrumb-nav"> <ol class="breadcrumb-nav-list"> <li class="breadcrumb-nav-list-item"><a href="/Intro2Bioinfo/">Intro to Bioinfo</a></li> <li class="breadcrumb-nav-list-item"><span>Lesson 2 - Basic Operations in Terminal and BASH</span></li> </ol> </nav> <div id="main-content" class="main-content"> <main> <h1 id="lesson-2-basic-operations-in-terminal-and-bash"> <a href="#lesson-2-basic-operations-in-terminal-and-bash" class="anchor-heading" aria-labelledby="lesson-2-basic-operations-in-terminal-and-bash"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Lesson 2: Basic Operations in Terminal and BASH </h1> <h2 id="auto-completion"> <a href="#auto-completion" class="anchor-heading" aria-labelledby="auto-completion"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Auto-completion </h2> <p>Let us return to our home directory <code class="language-plaintext highlighter-rouge">cd ~</code>. In this directory, create three folders named <code class="language-plaintext highlighter-rouge">test11</code>, <code class="language-plaintext highlighter-rouge">test20</code>, and <code class="language-plaintext highlighter-rouge">test21</code> using the <code class="language-plaintext highlighter-rouge">mkdir</code> command. Check your folder using <code class="language-plaintext highlighter-rouge">ls</code>.</p> <p>Afterwards, try typing <code class="language-plaintext highlighter-rouge">head te</code> and tap the <strong>tab button</strong> once. As you can see, it should auto-complete it to test as it is the next complete <code class="language-plaintext highlighter-rouge">string</code> without conflict in our folder. Now, that we have <code class="language-plaintext highlighter-rouge">head test</code> in our line, try tapping the <strong>tab button</strong> twice. It should ring a bell and prompt all possible matches with our initial user input.</p> <blockquote class="activity"> <p><code class="language-plaintext highlighter-rouge">test11 test20 test21</code></p> </blockquote> <p><br /></p> <h2 id="wildcards"> <a href="#wildcards" class="anchor-heading" aria-labelledby="wildcards"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Wildcards </h2> <p>Sometimes, we need to work on multiple files folders at the same time but they have different names (e.g. the test1, test 20 and test21). The default command-line shell that we are using allows manipulation of the files through the use of special characters called wildcards <code class="language-plaintext highlighter-rouge">*</code>.</p> <blockquote class="activity"> <p><strong><em><ins>Try running this command:</ins></em></strong></p> <p><code class="language-plaintext highlighter-rouge">ls te*</code></p> </blockquote> <p>The above command results in listing all files in the folder that starts with <code class="language-plaintext highlighter-rouge">te</code>. This can be used also in a more specific manner.</p> <blockquote class="activity"> <p><strong><em><ins>Try running this command:</ins></em></strong></p> <p><code class="language-plaintext highlighter-rouge">ls test*1</code></p> </blockquote> <p>Using the wildcard within the string matches anyword that starts with <code class="language-plaintext highlighter-rouge">test</code> and ends with <code class="language-plaintext highlighter-rouge">1</code>. It is useful also when runnning commands of files with the same file extension: <code class="language-plaintext highlighter-rouge">cp *.fasta ./folder/</code></p> <blockquote class="note"> <p>üìù <strong>Note:</strong> At the command line, the * means any character, any number of times (including 0 times).</p> </blockquote> <p>Let us try removing all the folder we made earlier that start with <code class="language-plaintext highlighter-rouge">test</code> simultaneously:</p> <blockquote class="activity"> <p><strong><em><ins>Try running this command:</ins></em></strong></p> <p><code class="language-plaintext highlighter-rouge">rm -r test*</code></p> </blockquote> <blockquote class="warning"> <p>‚ö†Ô∏è Removing files using wildcards should be done carefully as use of wildcards can result in unintended results (e.g. using <code class="language-plaintext highlighter-rouge">rm *</code> to delete all files within the folder)</p> </blockquote><hr /> <h3 id="the-question-mark"> <a href="#the-question-mark" class="anchor-heading" aria-labelledby="the-question-mark"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> <strong>The Question Mark</strong> </h3> <p>Another type of wildcard is the use of <code class="language-plaintext highlighter-rouge">?</code> mark. Unlike the <code class="language-plaintext highlighter-rouge">*</code> which can match 1 or more strings e.g. (<code class="language-plaintext highlighter-rouge">test11111</code> and <code class="language-plaintext highlighter-rouge">test11</code>), the use of <code class="language-plaintext highlighter-rouge">test*1</code> will match both files while use of <code class="language-plaintext highlighter-rouge">test?1</code> will only match <code class="language-plaintext highlighter-rouge">test11</code>.</p> <blockquote class="activity"> <p><strong><em><ins>Try running this command:</ins></em></strong></p> <p><code class="language-plaintext highlighter-rouge">ls test?1</code></p> </blockquote> <p>Which one matches your query?</p><hr /> <p><br /></p> <h2 id="redirectors"> <a href="#redirectors" class="anchor-heading" aria-labelledby="redirectors"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Redirectors </h2> <p>The redirection symbols allow the change of destination of outputs (oe even inputs in some instances). Rather than outputs directly showing to your terminal screen, it can be redirected to another command or file. Additionally, redirectors can also be use to push an input to a command.</p><hr /> <h3 id="the--and-"> <a href="#the--and-" class="anchor-heading" aria-labelledby="the--and-"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> <strong>The ‚Äò&gt;‚Äô and ‚Äò&lt;‚Äô</strong> </h3> <p>The ‚Äò&gt;‚Äô redirector indicates that the output of any command or process on its left will be transferred to a file indicated on its right. It‚Äôs most basic structure follows this:</p> <blockquote> <p><code class="language-plaintext highlighter-rouge">[command] [any arguments] &gt; [output file]</code></p> </blockquote> <blockquote class="activity"> <p><strong><em><ins>Try running this command:</ins></em></strong></p> <p><code class="language-plaintext highlighter-rouge">ls -all &gt; list.txt</code>, then open <code class="language-plaintext highlighter-rouge">list.txt</code></p> </blockquote> <blockquote class="warning"> <p>Be warned that using an <code class="language-plaintext highlighter-rouge">&gt;</code> output redirector towards an <strong>output file that already exists</strong> in the destination will lead to the <strong>overwriting</strong> of that file. To append an existing file (adding output on the end of the file), please use the append <code class="language-plaintext highlighter-rouge">&gt;&gt;</code> redirector instead.</p> </blockquote> <p>Meanwhile, the ‚Äò&lt;‚Äô redirector indicates that the file in the right of the <code class="language-plaintext highlighter-rouge">&lt;</code> will be used as an input to the command to its left: <code class="language-plaintext highlighter-rouge">[command][options/arguments] &lt; [input file]</code></p> <blockquote> <p><strong><em><ins>Example:</ins></em></strong></p> <p>```mail -s ‚ÄúNews Today‚Äù gccab@ymail.com &lt; NewsFlash.zip</p> </blockquote> <p>In this example, we can add the attachment <code class="language-plaintext highlighter-rouge">NewsFlash.zip</code> to the command line <code class="language-plaintext highlighter-rouge">mail</code> command rather than just sending plain text for email.</p> <blockquote class="info"> <p>The <code class="language-plaintext highlighter-rouge">mail</code> command is command-line command used to send email. It is useful to incorporate in programs or scripts that need immediate notification to the user if they have errors or the run have finished.</p> </blockquote><hr /> <h3 id="the-pipe-"> <a href="#the-pipe-" class="anchor-heading" aria-labelledby="the-pipe-"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> <strong>The pipe ‚Äò|‚Äô</strong> </h3> <p>If you want your output to be redirected not into a file but to another process or command, the <code class="language-plaintext highlighter-rouge">|</code> or pipe redirector is primarily used. It can be used not only once on a line but can be used to chain several commands to gether. A basic structure would look like this:</p> <blockquote> <p><code class="language-plaintext highlighter-rouge">[command 1] | [command 2] | [command 3] &gt; [output file or STDOUT]</code></p> </blockquote> <blockquote class="activity"> <p><strong><em><ins>Try running this command:</ins></em></strong></p> <p><code class="language-plaintext highlighter-rouge">ls ./data/all | head -n 20 &gt; list.txt</code></p> </blockquote><hr /> <h3 id="error-redirectors"> <a href="#error-redirectors" class="anchor-heading" aria-labelledby="error-redirectors"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> <strong>Error redirectors</strong> </h3> <p>Several programs often outputed also the errors in STDOUT or terminal screen. You can separate the errors and redirect it to a new distinct file using <code class="language-plaintext highlighter-rouge">2&gt;</code> redirection. It works similar as the normal output redirection but the difference is it will only filter ‚Äòerror‚Äô outputs designated by the program you use.</p> <p>The ‚Äòerror‚Äô redirector <code class="language-plaintext highlighter-rouge">2&gt;</code> can be used alongside a ‚Äòoutput‚Äô redirector but designate it as such <code class="language-plaintext highlighter-rouge">1&gt;</code>. One example would be like this: <code class="language-plaintext highlighter-rouge">head nonexistentfile.txt 1&gt; file.out 2&gt; file.err</code>.</p> <p><br /></p> <h1 id="bioinformatic-file-types-and-formats"> <a href="#bioinformatic-file-types-and-formats" class="anchor-heading" aria-labelledby="bioinformatic-file-types-and-formats"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Bioinformatic File types and formats </h1> <h2 id="sequence-file-formats"> <a href="#sequence-file-formats" class="anchor-heading" aria-labelledby="sequence-file-formats"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Sequence File Formats </h2> <p>Sequencing of DNA and RNA often produce signals of light, chromatogram peaks or electric current as an indication of a nucleotide base. These raw signals are converted into ASCII or computer friendly code using a ‚Äòbasecaller‚Äô or basecalling program. Most of the files that we analysed in bioinformatic analysis are already in this ASCII or alphanumeric format that is readable to both machines and humans.</p> <p class="info">To learn more about basecaller, please look into in this <a href="https://nanoporetech.com/platform/technology/basecalling">nanopore guide</a>.</p> <h2 id="the--fastq-format"> <a href="#the--fastq-format" class="anchor-heading" aria-labelledby="the--fastq-format"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> The FASTQ format </h2> <p>The most common file format after basecalling is the <strong><em>fastq</em></strong> format or colloquially known as <em>Quality FASTA</em> format. The fastq files contained both nucleotide sequence information, represented with characters as <code class="language-plaintext highlighter-rouge">AGCTN*</code>, and the quality score of each nucleotide, which are represented by different alphanumeric characters. The quality score is based on <strong>Phred Q Score</strong> system in which quality is computed as probabilty of error <em>Pe</em> [2], as computed by this equation:</p> <blockquote class="activity"> <p><code class="language-plaintext highlighter-rouge">Qphred=-10 x log10(Pe)</code></p> </blockquote> <p>An example of a <strong>fastq</strong> file:</p> <div class="image3"> <img src="https://biocorecrg.github.io/RNAseq_course_2019/images/fastq_format.png" alt="Sourced from https://biocorecrg.github.io/RNAseq_course_2019/fastq.html" /> </div> <blockquote class="highlight"> <p>The FASTQ format consists of four lines:</p> <ol> <li><strong>Header</strong> ‚Äî Starts with <code class="language-plaintext highlighter-rouge">@</code>, followed by a sequence ID and metadata.</li> <li><strong>Sequence</strong> ‚Äî The raw nucleotide sequence (e.g., <code class="language-plaintext highlighter-rouge">AGCT...</code>).</li> <li><strong>Separator</strong> ‚Äî A <code class="language-plaintext highlighter-rouge">+</code> symbol separating the sequence and quality.</li> <li><strong>Quality</strong> ‚Äî ASCII-encoded Phred quality scores.</li> </ol> </blockquote> <h3 id="phred-score"> <a href="#phred-score" class="anchor-heading" aria-labelledby="phred-score"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Phred Score </h3> <p>The quality scores ranges from 0-40 with 40 indicating the highest level of quality. Phred quality score can be interpreted as a factor of accuracy as shown in this table:</p> <center>Quality Scores and Accuracy</center> <div class="table-wrapper"><table> <thead> <tr> <th>Phred Score</th> <th>Probability of incorrect call</th> <th>Base Call Accuracy</th> </tr> </thead> <tbody> <tr> <td>10</td> <td>1 in 10</td> <td>90%</td> </tr> <tr> <td>20</td> <td>1 in 100</td> <td>99%</td> </tr> <tr> <td>30</td> <td>1 in 1,000</td> <td>99.9%</td> </tr> <tr> <td>40</td> <td>1 in 10,000</td> <td>99.99%</td> </tr> </tbody> </table></div> <p>The ASCII representation of this quality score in fastq files can be seen in this image adapted from <a href="https://learn.gencore.bio.nyu.edu/ngs-file-formats/quality-scores/">NYU</a>: <br /></p> <div class="image"> <img src="https://learn.gencore.bio.nyu.edu/wp-content/uploads/2018/01/Screen-Shot-2018-01-07-at-1.34.33-PM-768x492.png" /> </div> <p><br /></p> <p class="note">Phred-64 format is only used by Illumina sequencers while Phred-33 is primarily used by other sequencing file format.</p> <h2 id="the--fasta-format"> <a href="#the--fasta-format" class="anchor-heading" aria-labelledby="the--fasta-format"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> The FASTA format </h2> <p>FASTA files are the most common file format you will encounter when doing bioinformatic analysis. Unlike FASTQ files with quality score, FASTA files primarily contains only the sequence information using <code class="language-plaintext highlighter-rouge">AGCT</code> characters for nucleotides. Aside from this, the <code class="language-plaintext highlighter-rouge">N</code> and <code class="language-plaintext highlighter-rouge">*</code> are also used as unspecified based and stop character.</p> <p><br /></p> <div class="image3"> <center> <img src="https://www.researchgate.net/profile/Jiangning-Song/publication/331702618/figure/fig2/AS:745823247810566@1554829527399/An-example-of-the-FASTA-format-used-in-iLearn.ppm" /> </center> </div> <center><ins>A nucleotide FASTA File</ins></center> <p><br /></p> <p>The <strong><em>FASTA</em></strong> file format is composed of two lines per sequence. The first line starts with <code class="language-plaintext highlighter-rouge">&gt;</code> composes the header line. It often has the sequence name and other properties depending on the program used. The second line is composed of the nucleotide sequence file. This can be on capital or small characters.</p> <p>When reading FASTA format of protein sequences, we use single amino acid representations.</p> <div class="image3"> <center> <img src="https://www.integratedbreeding.net/courses/genomics-and-comparative-genomics/www.generationcp.org/UserFiles/Image/genomics/amino.gif" /> </center> </div> <p>Here is a sample of a protein FASTA file: <br /></p> <div class="image"> <center> <img src="https://training.galaxyproject.org/training-material/topics/proteomics/images/Fasta_sequence.png" /> </center> <center><ins>Protein FASTA File</ins></center> </div> <blockquote class="warning"> <p>The two formats should not be used at the same file as the nucleotide fasta format can be interpreted as a protein fasta file.</p> </blockquote> <p><br /></p> <h2 id="alignment-format"> <a href="#alignment-format" class="anchor-heading" aria-labelledby="alignment-format"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Alignment Format </h2> <p>In many downstream analysis, fasta/fastq files are aligned to create a large genome sequence file. Among the most common alignment format used here are the <strong>SAM</strong> and <strong>BAM</strong> format. ‚ÄòBAM‚Äô is the binary file format of ‚ÄòSAM‚Äô and often used for its smaller file size.</p> <h3 id="the-sam-or-sequence-alignment-map-format"> <a href="#the-sam-or-sequence-alignment-map-format" class="anchor-heading" aria-labelledby="the-sam-or-sequence-alignment-map-format"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> The SAM or Sequence Alignment Map format </h3> <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="o">%%</span><span class="n">script</span> <span class="n">false</span> <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="k">raise</span><span class="o">-</span><span class="n">error</span>
<span class="mi">1</span><span class="p">:</span><span class="mi">497</span><span class="p">:</span><span class="n">R</span><span class="p">:</span><span class="o">-</span><span class="mi">272</span><span class="o">+</span><span class="mi">13</span><span class="n">M17D24M</span>	<span class="mi">113</span>	<span class="mi">1</span>	<span class="mi">497</span>	<span class="mi">37</span>	<span class="mi">37</span><span class="n">M</span>	<span class="mi">15</span>	<span class="mi">100338662</span>	<span class="mi">0</span>	<span class="n">CGGGTCTGACCTGAGGAGAACTGTGCTCCGCCTTCAG</span>	<span class="mi">0</span><span class="p">;</span><span class="o">==-==</span><span class="mi">9</span><span class="p">;</span><span class="o">&gt;&gt;&gt;&gt;&gt;=&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;=&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;</span>	<span class="n">XT</span><span class="p">:</span><span class="n">A</span><span class="p">:</span><span class="n">U</span>	<span class="n">NM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">SM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">37</span>	<span class="n">AM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">X0</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">1</span>	<span class="n">X1</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">XM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">XO</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">XG</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">MD</span><span class="p">:</span><span class="n">Z</span><span class="p">:</span><span class="mi">37</span>
<span class="mi">19</span><span class="p">:</span><span class="mi">20389</span><span class="p">:</span><span class="n">F</span><span class="p">:</span><span class="mi">275</span><span class="o">+</span><span class="mi">18</span><span class="n">M2D19M</span>	<span class="mi">99</span>	<span class="mi">1</span>	<span class="mi">17644</span>	<span class="mi">0</span>	<span class="mi">37</span><span class="n">M</span>	<span class="o">=</span>	<span class="mi">17919</span>	<span class="mi">314</span>	<span class="n">TATGACTGCTAATAATACCTACACATGTTAGAACCAT</span>	<span class="o">&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&lt;&lt;&gt;&gt;&gt;&lt;&lt;&gt;&gt;</span><span class="mi">4</span><span class="p">::</span><span class="o">&gt;&gt;</span><span class="p">:</span><span class="o">&lt;</span><span class="mi">9</span>	<span class="n">RG</span><span class="p">:</span><span class="n">Z</span><span class="p">:</span><span class="n">UM0098</span><span class="p">:</span><span class="mi">1</span>	<span class="n">XT</span><span class="p">:</span><span class="n">A</span><span class="p">:</span><span class="n">R</span>	<span class="n">NM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">SM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">AM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">X0</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">4</span>	<span class="n">X1</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">XM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">XO</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">XG</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">MD</span><span class="p">:</span><span class="n">Z</span><span class="p">:</span><span class="mi">37</span>
<span class="mi">19</span><span class="p">:</span><span class="mi">20389</span><span class="p">:</span><span class="n">F</span><span class="p">:</span><span class="mi">275</span><span class="o">+</span><span class="mi">18</span><span class="n">M2D19M</span>	<span class="mi">147</span>	<span class="mi">1</span>	<span class="mi">17919</span>	<span class="mi">0</span>	<span class="mi">18</span><span class="n">M2D19M</span>	<span class="o">=</span>	<span class="mi">17644</span>	<span class="o">-</span><span class="mi">314</span>	<span class="n">GTAGTACCAACTGTAAGTCCTTATCTTCATACTTTGT</span>	<span class="p">;</span><span class="mi">44999</span><span class="p">;</span><span class="mi">499</span><span class="o">&lt;</span><span class="mi">8</span><span class="o">&lt;</span><span class="mi">8</span><span class="o">&lt;&lt;&lt;</span><span class="mi">8</span><span class="o">&lt;&lt;&gt;&lt;&lt;&lt;&lt;&gt;&lt;</span><span class="mi">7</span><span class="o">&lt;</span><span class="p">;</span><span class="o">&lt;&lt;&lt;&gt;&gt;&lt;&lt;</span>	<span class="n">XT</span><span class="p">:</span><span class="n">A</span><span class="p">:</span><span class="n">R</span>	<span class="n">NM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">2</span>	<span class="n">SM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">AM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">X0</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">4</span>	<span class="n">X1</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">XM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">XO</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">1</span>	<span class="n">XG</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">2</span>	<span class="n">MD</span><span class="p">:</span><span class="n">Z</span><span class="p">:</span><span class="mi">18</span><span class="o">^</span><span class="n">CA19</span>
<span class="mi">9</span><span class="p">:</span><span class="mi">21597</span><span class="o">+</span><span class="mi">10</span><span class="n">M2I25M</span><span class="p">:</span><span class="n">R</span><span class="p">:</span><span class="o">-</span><span class="mi">209</span>	<span class="mi">83</span>	<span class="mi">1</span>	<span class="mi">21678</span>	<span class="mi">0</span>	<span class="mi">8</span><span class="n">M2I27M</span>	<span class="o">=</span>	<span class="mi">21469</span>	<span class="o">-</span><span class="mi">244</span>	<span class="n">CACCACATCACATATACCAAGCCTGGCTGTGTCTTCT</span>	<span class="o">&lt;</span><span class="p">;</span><span class="mi">9</span><span class="o">&lt;&lt;</span><span class="mi">5</span><span class="o">&gt;&lt;&lt;&lt;&lt;&gt;&lt;&lt;&lt;&gt;&gt;&lt;&lt;&gt;&lt;&gt;&gt;&lt;</span><span class="mi">9</span><span class="o">&gt;&gt;&lt;&gt;&gt;&gt;</span><span class="mi">9</span><span class="o">&gt;&gt;&gt;&lt;&gt;</span>	<span class="n">XT</span><span class="p">:</span><span class="n">A</span><span class="p">:</span><span class="n">R</span>	<span class="n">NM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">2</span>	<span class="n">SM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">AM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">X0</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">5</span>	<span class="n">X1</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">XM</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>	<span class="n">XO</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">1</span>	<span class="n">XG</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">2</span>	<span class="n">MD</span><span class="p">:</span><span class="n">Z</span><span class="p">:</span><span class="mi">35</span>

</code></pre></div></div> <p>As shown above, it is composed of several columns that describes a reads position in the genome. The meaning of each column can be seen here:</p> <p><img src="https://bioinformatics.media.uconn.edu/wp-content/uploads/sites/15/2017/01/format2-768x247.png" /></p> <p>We will not discuss this in detail but will tackle in the future.</p> <p><br /></p> <h2 id="other-file-formats"> <a href="#other-file-formats" class="anchor-heading" aria-labelledby="other-file-formats"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Other File formats </h2> <p>Multiple other file formats are used in Bioinformatics. We will tackle them in detail when we encounter them in our analysis. Here are some of the file formats that are common in a bioinformatic workflow.</p> <center><em><b>Bioinformatic files</b></em></center> <div class="table-wrapper"><table> <thead> <tr> <th><strong>File Formats</strong></th> <th><strong>Extension</strong></th> <th><strong>Definitions</strong></th> </tr> </thead> <tbody> <tr> <td>FASTQ</td> <td>.fastq or .fq</td> <td>contains sequence information and the quality of the sequences</td> </tr> <tr> <td>FASTA</td> <td>.fasta or .fa</td> <td>basic file containing description and sequence information</td> </tr> <tr> <td>¬†</td> <td>.fna</td> <td>an alternative extension used to denote a nucleotide fasta format</td> </tr> <tr> <td>¬†</td> <td>.faa</td> <td>an alternative extension used to denote a protein fasta format</td> </tr> <tr> <td>¬†</td> <td>.fa.gz or .fasta.gz or .fq.gz</td> <td>gunzipped compressed format of FASTA or FASTQ for smaller file size</td> </tr> <tr> <td>SAM</td> <td>.sam</td> <td>a sequence alignment format that details position of each nucleotide. May include quality score and other options depending on the program used</td> </tr> <tr> <td>BAM</td> <td>.bam</td> <td>a compressed or binary form of SAM to lower file size</td> </tr> <tr> <td>General Feature Format</td> <td>.gff or .gff2 or .gff3</td> <td>is a file format used to describe genes and other general feature of the genome</td> </tr> <tr> <td>¬†</td> <td>¬†</td> <td>It is consisting of several columns describing the genes, gene names, position and orientation of the genes</td> </tr> <tr> <td>Gene Transfer format</td> <td>.gtf</td> <td>Similar to GFF files but may containe gene structure data</td> </tr> <tr> <td>Variant Calling File</td> <td>.vcf</td> <td>VCf is the standard format when storing gene variation data such as <em>indels</em> and <em>single nucleotide polymorphisms (SNPs)</em></td> </tr> </tbody> </table></div> <p><br /></p> <h1 id="summary"> <a href="#summary" class="anchor-heading" aria-labelledby="summary"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Summary </h1> <h2 id="common-operations-in-terminal"> <a href="#common-operations-in-terminal" class="anchor-heading" aria-labelledby="common-operations-in-terminal"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Common operations in Terminal </h2> <div class="table-wrapper"><table> <thead> <tr> <th><strong>Terminologies</strong></th> <th><strong>Definitions</strong></th> </tr> </thead> <tbody> <tr> <td><code class="language-plaintext highlighter-rouge">tab</code></td> <td>Auto-completion of files or programs available in the terminal</td> </tr> <tr> <td><code class="language-plaintext highlighter-rouge">*</code></td> <td>A catch-all wildcard that matches 0 or more instances</td> </tr> <tr> <td><code class="language-plaintext highlighter-rouge">?</code></td> <td>A single character wildcard that matches just 1 character instance</td> </tr> <tr> <td><code class="language-plaintext highlighter-rouge">&gt;</code></td> <td>Output redirector that allows both STDOUT and STDERR to be save to a file</td> </tr> <tr> <td><code class="language-plaintext highlighter-rouge">&lt;</code></td> <td>Input redirector that allows files to fill in as STDIN to the command or program</td> </tr> <tr> <td><code class="language-plaintext highlighter-rouge">&gt;&gt;</code></td> <td>Append redirector add STDOUT to the end of the file rather than replacing it</td> </tr> <tr> <td><code class="language-plaintext highlighter-rouge">2&gt;</code></td> <td>error redirection separates STDERR outputs to a file</td> </tr> <tr> <td><code class="language-plaintext highlighter-rouge">1&gt;</code></td> <td>STDOUT only redirection to a file</td> </tr> <tr> <td><code class="language-plaintext highlighter-rouge">\|</code></td> <td>Pipe redirection allows to chain a STDOUT as STDIN to another program</td> </tr> <tr> <td>STDIN</td> <td>Standard input are inputs that are often provided to a program or command via terminal</td> </tr> <tr> <td>STDOUT</td> <td>Standard output are output of programs and commands and are often shown into the terminal rather than a file</td> </tr> <tr> <td>STDERR</td> <td>Standard errors are error outputs of programs and commands that is also often shown in terminal</td> </tr> </tbody> </table></div> <h1 id="acknowledgement"> <a href="#acknowledgement" class="anchor-heading" aria-labelledby="acknowledgement"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Acknowledgement </h1> <p>This tutorial is adapted from <em>Intro to Unix</em> from Happy Belly Bioinformatics by Michael D. Lee or known as <a href="https://astrobiomike.github.io/unix/">AstroBioMike in github</a>.</p> <h1 id="citation"> <a href="#citation" class="anchor-heading" aria-labelledby="citation"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Citation </h1> <p>[1] Lee, M. (2019). Happy Belly Bioinformatics: an open-source resource dedicated to helping biologists utilize bioinformatics. Journal of Open Source Education, 4(41), 53, https://doi.org/10.21105/jose.00053</p> <p>[2] Cock, P.J.A, C.J. Fields, N. Goto, M.L. Heuer, &amp; P.M. Rice (2010). The Sanger FASTQ file format for sequences with quality scores, and the Solexa/Illumina FASTQ variants. Nucleic Acids Research, 38(6):1767-1771, https://doi.org/10.1093%2Fnar%2Fgkp1137</p> </main> </div> </div> <div class="search-overlay"></div> </div> </body> </html>
