<!DOCTYPE html> <html lang="en-US"> <head> <meta charset="UTF-8"> <meta http-equiv="X-UA-Compatible" content="IE=Edge"> <link rel="stylesheet" href="/assets/css/just-the-docs-default.css"> <link rel="stylesheet" href="/assets/css/just-the-docs-head-nav.css" id="jtd-head-nav-stylesheet"> <style id="jtd-nav-activation"> .site-nav > ul.nav-list:first-child > li > a, .site-nav > ul.nav-list:first-child > li > ul > li:not(:nth-child(3)) > a, .site-nav > ul.nav-list:first-child > li > ul > li > ul > li a { background-image: none; } .site-nav > ul.nav-list:not(:first-child) a, .site-nav li.external a { background-image: none; } .site-nav > ul.nav-list:first-child > li:nth-child(3) > ul > li:nth-child(3) > a { font-weight: 600; text-decoration: none; }.site-nav > ul.nav-list:first-child > li:nth-child(3) > button svg, .site-nav > ul.nav-list:first-child > li:nth-child(3) > ul > li:nth-child(3) > button svg { transform: rotate(-90deg); }.site-nav > ul.nav-list:first-child > li.nav-list-item:nth-child(3) > ul.nav-list, .site-nav > ul.nav-list:first-child > li.nav-list-item:nth-child(3) > ul.nav-list > li.nav-list-item:nth-child(3) > ul.nav-list { display: block; } </style> <script src="/assets/js/vendor/lunr.min.js"></script> <script src="/assets/js/just-the-docs.js"></script> <meta name="viewport" content="width=device-width, initial-scale=1"> <!-- Begin Jekyll SEO tag v2.8.0 --> <title>Lesson 3 | üß¨ MEB Training Suite</title> <meta name="generator" content="Jekyll v4.3.4" /> <meta property="og:title" content="Lesson 3" /> <meta property="og:locale" content="en_US" /> <meta name="description" content="Bookmark this to keep an eye on my project updates!" /> <meta property="og:description" content="Bookmark this to keep an eye on my project updates!" /> <link rel="canonical" href="http://localhost:4000/Intro2Bioinfo/Lesson3.html" /> <meta property="og:url" content="http://localhost:4000/Intro2Bioinfo/Lesson3.html" /> <meta property="og:site_name" content="üß¨ MEB Training Suite" /> <meta property="og:type" content="website" /> <meta name="twitter:card" content="summary" /> <meta property="twitter:title" content="Lesson 3" /> <script type="application/ld+json"> {"@context":"https://schema.org","@type":"WebPage","description":"Bookmark this to keep an eye on my project updates!","headline":"Lesson 3","url":"http://localhost:4000/Intro2Bioinfo/Lesson3.html"}</script> <!-- End Jekyll SEO tag --> </head> <body> <a class="skip-to-main" href="#main-content">Skip to main content</a> <svg xmlns="http://www.w3.org/2000/svg" class="d-none"> <symbol id="svg-link" viewBox="0 0 24 24"> <title>Link</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-link"> <path d="M10 13a5 5 0 0 0 7.54.54l3-3a5 5 0 0 0-7.07-7.07l-1.72 1.71"></path><path d="M14 11a5 5 0 0 0-7.54-.54l-3 3a5 5 0 0 0 7.07 7.07l1.71-1.71"></path> </svg> </symbol> <symbol id="svg-menu" viewBox="0 0 24 24"> <title>Menu</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-menu"> <line x1="3" y1="12" x2="21" y2="12"></line><line x1="3" y1="6" x2="21" y2="6"></line><line x1="3" y1="18" x2="21" y2="18"></line> </svg> </symbol> <symbol id="svg-arrow-right" viewBox="0 0 24 24"> <title>Expand</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-chevron-right"> <polyline points="9 18 15 12 9 6"></polyline> </svg> </symbol> <!-- Feather. MIT License: https://github.com/feathericons/feather/blob/master/LICENSE --> <symbol id="svg-external-link" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-external-link"> <title id="svg-external-link-title">(external link)</title> <path d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"></path><polyline points="15 3 21 3 21 9"></polyline><line x1="10" y1="14" x2="21" y2="3"></line> </symbol> <symbol id="svg-doc" viewBox="0 0 24 24"> <title>Document</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-file"> <path d="M13 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V9z"></path><polyline points="13 2 13 9 20 9"></polyline> </svg> </symbol> <symbol id="svg-search" viewBox="0 0 24 24"> <title>Search</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-search"> <circle cx="11" cy="11" r="8"></circle><line x1="21" y1="21" x2="16.65" y2="16.65"></line> </svg> </symbol> <!-- Bootstrap Icons. MIT License: https://github.com/twbs/icons/blob/main/LICENSE.md --> <symbol id="svg-copy" viewBox="0 0 16 16"> <title>Copy</title> <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-clipboard" viewBox="0 0 16 16"> <path d="M4 1.5H3a2 2 0 0 0-2 2V14a2 2 0 0 0 2 2h10a2 2 0 0 0 2-2V3.5a2 2 0 0 0-2-2h-1v1h1a1 1 0 0 1 1 1V14a1 1 0 0 1-1 1H3a1 1 0 0 1-1-1V3.5a1 1 0 0 1 1-1h1v-1z"/> <path d="M9.5 1a.5.5 0 0 1 .5.5v1a.5.5 0 0 1-.5.5h-3a.5.5 0 0 1-.5-.5v-1a.5.5 0 0 1 .5-.5h3zm-3-1A1.5 1.5 0 0 0 5 1.5v1A1.5 1.5 0 0 0 6.5 4h3A1.5 1.5 0 0 0 11 2.5v-1A1.5 1.5 0 0 0 9.5 0h-3z"/> </svg> </symbol> <symbol id="svg-copied" viewBox="0 0 16 16"> <title>Copied</title> <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-clipboard-check-fill" viewBox="0 0 16 16"> <path d="M6.5 0A1.5 1.5 0 0 0 5 1.5v1A1.5 1.5 0 0 0 6.5 4h3A1.5 1.5 0 0 0 11 2.5v-1A1.5 1.5 0 0 0 9.5 0h-3Zm3 1a.5.5 0 0 1 .5.5v1a.5.5 0 0 1-.5.5h-3a.5.5 0 0 1-.5-.5v-1a.5.5 0 0 1 .5-.5h3Z"/> <path d="M4 1.5H3a2 2 0 0 0-2 2V14a2 2 0 0 0 2 2h10a2 2 0 0 0 2-2V3.5a2 2 0 0 0-2-2h-1v1A2.5 2.5 0 0 1 9.5 5h-3A2.5 2.5 0 0 1 4 2.5v-1Zm6.854 7.354-3 3a.5.5 0 0 1-.708 0l-1.5-1.5a.5.5 0 0 1 .708-.708L7.5 10.793l2.646-2.647a.5.5 0 0 1 .708.708Z"/> </svg> </symbol> </svg> <div class="side-bar"> <div class="site-header" role="banner"> <a href="/" class="site-title lh-tight"> üß¨ MEB Training Suite </a> <button id="menu-button" class="site-button btn-reset" aria-label="Toggle menu" aria-pressed="false"> <svg viewBox="0 0 24 24" class="icon" aria-hidden="true"><use xlink:href="#svg-menu"></use></svg> </button> </div> <nav aria-label="Main" id="site-nav" class="site-nav"> <ul class="nav-list"><li class="nav-list-item"><a href="/" class="nav-list-link">Welcome</a></li><li class="nav-list-item"><a href="/getting-started.html" class="nav-list-link">Getting Started</a></li><li class="nav-list-item"><button class="nav-list-expander btn-reset" aria-label="toggle items in Intro to Bioinfo category" aria-pressed="false"> <svg viewBox="0 0 24 24" aria-hidden="true"><use xlink:href="#svg-arrow-right"></use></svg> </button><a href="/Intro2Bioinfo/" class="nav-list-link">Intro to Bioinfo</a><ul class="nav-list"><li class="nav-list-item"><a href="/Intro2Bioinfo/Lesson1.html" class="nav-list-link">Lesson 1</a></li><li class="nav-list-item"><a href="/Intro2Bioinfo/Lesson2.html" class="nav-list-link">Lesson 2</a></li><li class="nav-list-item"><a href="/Intro2Bioinfo/Lesson3.html" class="nav-list-link">Lesson 3</a></li></ul></li></ul> </nav> <footer class="site-footer"> This site uses <a href="https://github.com/just-the-docs/just-the-docs">Just the Docs</a>, a documentation theme for Jekyll. </footer> </div> <div class="main" id="top"> <div id="main-header" class="main-header"> <div class="search" role="search"> <div class="search-input-wrap"> <input type="text" id="search-input" class="search-input" tabindex="0" placeholder="Search üß¨ MEB Training Suite" aria-label="Search üß¨ MEB Training Suite" autocomplete="off"> <label for="search-input" class="search-label"><svg viewBox="0 0 24 24" class="search-icon"><use xlink:href="#svg-search"></use></svg></label> </div> <div id="search-results" class="search-results"></div> </div> <nav aria-label="Auxiliary" class="aux-nav"> <ul class="aux-nav-list"> <li class="aux-nav-list-item"> <a href="https://github.com/gamalielcabria" class="site-button" > Template Repository </a> </li> </ul> </nav> </div> <div class="main-content-wrap"> <nav aria-label="Breadcrumb" class="breadcrumb-nav"> <ol class="breadcrumb-nav-list"> <li class="breadcrumb-nav-list-item"><a href="/Intro2Bioinfo/">Intro to Bioinfo</a></li> <li class="breadcrumb-nav-list-item"><span>Lesson 3</span></li> </ol> </nav> <div id="main-content" class="main-content"> <main> <h1 id="lesson-3-the-commands-to-rule-them-all"> <a href="#lesson-3-the-commands-to-rule-them-all" class="anchor-heading" aria-labelledby="lesson-3-the-commands-to-rule-them-all"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Lesson 3: The commands to rule them all </h1> <h2 id="basic-bioinformatics-commands"> <a href="#basic-bioinformatics-commands" class="anchor-heading" aria-labelledby="basic-bioinformatics-commands"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Basic bioinformatics commands: </h2> <p>There are tons of built-in commands within Linux OSes that are very useful for bioinformatic analysis. In this section, we will examine a few of the most commonly used commands in bioinformatics. We will adapt content from the <a href="https://astrobiomike.github.io/unix/six-glorious-commands">Happy Belly Bioinformatics</a>.</p> <blockquote class="note"> <p>The following lesson and activity should be tried using a <strong>linux terminal</strong>. Please use your WSL2 and/or register an account to <a href="https://webminal.org/">Webminal</a>.</p> </blockquote><hr /> <h2 id="history-history-n"> <a href="#history-history-n" class="anchor-heading" aria-labelledby="history-history-n"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> History: <code class="language-plaintext highlighter-rouge">history &lt;n&gt;</code> </h2> <p>Look into the recent commands run in the terminal for the past <code class="language-plaintext highlighter-rouge">n</code> lines. Useful for backtracking the commands you have previously run.</p> <blockquote class="activity"> <p><strong><ins>Try this command:</ins></strong></p> <p><code class="language-plaintext highlighter-rouge">history 20</code></p> </blockquote> <p>This should display the last 20 files you have entered in the command-line. If you are searching a specific command, you can also combine <code class="language-plaintext highlighter-rouge">history</code> with <code class="language-plaintext highlighter-rouge">grep</code>(see below) to search a specific command you‚Äôve previously used.</p> <blockquote class="activity"> <p><strong><ins>Try this command:</ins></strong></p> <p><code class="language-plaintext highlighter-rouge">history | grep "ls"</code></p> </blockquote> <p>It should be noted that you can recall also previous commands using your up ‚¨ÜÔ∏è and down ‚¨áÔ∏è arrow key.</p><hr /> <h2 id="gnu-wget-wget-url"> <a href="#gnu-wget-wget-url" class="anchor-heading" aria-labelledby="gnu-wget-wget-url"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> GNU Wget <code class="language-plaintext highlighter-rouge">wget &lt;url&gt;</code> </h2> <p><code class="language-plaintext highlighter-rouge">wget</code> is a command-line utility for downloading files from the web. It supports HTTP, HTTPS, and FTP protocols and is useful for retrieving datasets, software packages, or scripts directly into your working directory.</p> <blockquote class="activity"> <p><strong><ins>Try this command:</ins></strong></p> <p><code class="language-plaintext highlighter-rouge">wget https://github.com/gamalielcabria/gamalielcabria.github.io/blob/main/Intro2Bioinfo/files/gene_counts.tsv</code></p> </blockquote> <p>The previous command will download the <code class="language-plaintext highlighter-rouge">gene_counts.tsv</code> directly into your current working directory. If you want it to be located in a <strong>different directory</strong>, you can use the command:</p> <p><code class="language-plaintext highlighter-rouge">wget &lt;url&gt; -P &lt;destination path&gt;</code>.</p> <p>To download a folder with multiple files inside, it would be suggested to zip the files or if not possible do a recursive download:</p> <blockquote class="activity"> <p><strong><ins>Try this command:</ins></strong></p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Create a destination folder
mkdir -p ./destination

#download recursively to the destination folder
wget -r -np -nH --cut-dirs=5 -A "*.gtf.gz" ftp://ftp.ensembl.org/pub/release-110/gtf/homo_sapiens/ -P ./destination
</code></pre></div> </div> <blockquote class="info"> <p>This command will download all <code class="language-plaintext highlighter-rouge">*gtf.gz</code> files from ENSEMBL</p> <ul> <li><code class="language-plaintext highlighter-rouge">-r</code>: Recursive download</li> <li><code class="language-plaintext highlighter-rouge">-np</code>: No parent: don‚Äôt go to parent directories</li> <li><code class="language-plaintext highlighter-rouge">-nH</code>: No host directory: avoid creating <code class="language-plaintext highlighter-rouge">ftp.ensembl.org/</code></li> <li><code class="language-plaintext highlighter-rouge">--cut-dirs=5</code>: Skip 5 directory levels (so files save cleanly)</li> <li><code class="language-plaintext highlighter-rouge">-A "*.gtf.gz"</code>: Only download files ending in <code class="language-plaintext highlighter-rouge">.gtf.gz</code></li> <li><code class="language-plaintext highlighter-rouge">-P</code>: Destination folder of the file</li> </ul> </blockquote> </blockquote><hr /> <h2 id="gunzip-compression-gzip--d-k-file"> <a href="#gunzip-compression-gzip--d-k-file" class="anchor-heading" aria-labelledby="gunzip-compression-gzip--d-k-file"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Gunzip Compression <code class="language-plaintext highlighter-rouge">gzip [-d|-k] &lt;file&gt;</code> </h2> <p><code class="language-plaintext highlighter-rouge">gzip</code> is a widely used command-line tool for compressing text files, such as FASTA, FASTQ, GTF, or VCF files, into <code class="language-plaintext highlighter-rouge">.gz</code> format. This is especially useful in bioinformatics, where raw data files can be very large.</p> <p>This command compresses data.txt into data.txt.gz and removes the original file by default.</p> <blockquote class="activity"> <p><code class="language-plaintext highlighter-rouge">gzip data.txt</code></p> </blockquote> <p>The plaing <code class="language-plaintext highlighter-rouge">gzip</code> command will remove the original file and be replaced with <code class="language-plaintext highlighter-rouge">data.txt.gz</code>. Use <code class="language-plaintext highlighter-rouge">-k</code> to keep the original file:</p> <blockquote class="activity"> <p><code class="language-plaintext highlighter-rouge">gzip -k data.txt</code></p> </blockquote> <p>Meanwhile, decompressing a file uses <code class="language-plaintext highlighter-rouge">-d</code> option or <code class="language-plaintext highlighter-rouge">gunzip</code> command.</p> <blockquote class="activity"> <p><code class="language-plaintext highlighter-rouge">gzip -d data.txt.gz</code></p> </blockquote> <p>To retain the original <code class="language-plaintext highlighter-rouge">gz</code> file:</p> <blockquote class="activity"> <p><code class="language-plaintext highlighter-rouge">gunzip -c data.txt.gz &gt; data.txt</code></p> </blockquote> <p>Compare the sizes of the files between the compressed and uncompressed files. Did you notice any difference?</p> <div class="table-wrapper"><table> <thead> <tr> <th>File Type</th> <th>Compression Ratio</th> <th>Notes</th> </tr> </thead> <tbody> <tr> <td>FASTA</td> <td>2‚Äì4√ó smaller</td> <td>Compresses well due to repetitive sequences</td> </tr> <tr> <td>FASTQ</td> <td>3‚Äì6√ó smaller</td> <td>Especially efficient for raw reads</td> </tr> <tr> <td>GTF/GFF</td> <td>3‚Äì5√ó smaller</td> <td>Mostly text-based, highly compressible</td> </tr> <tr> <td>VCF</td> <td>4‚Äì8√ó smaller</td> <td>Works well but block compression (<code class="language-plaintext highlighter-rouge">bgzip</code>) often preferred</td> </tr> </tbody> </table></div> <blockquote class="notice"> <p>Gunzip compressed files cannot be read by normal text editor or text file view commands such as <code class="language-plaintext highlighter-rouge">more</code>, <code class="language-plaintext highlighter-rouge">cat</code>, and <code class="language-plaintext highlighter-rouge">nano</code>. They will appear gibberish because they are binary. However, <code class="language-plaintext highlighter-rouge">less</code> command are able to interpret them.</p> </blockquote> <blockquote class="note"> <p><strong>zcat</strong> <code class="language-plaintext highlighter-rouge">zcat &lt;gzipped file&gt;</code></p> <p>Given that concatenate <code class="language-plaintext highlighter-rouge">cat</code> command does not work with gzipped files. The <code class="language-plaintext highlighter-rouge">zcat</code> command replaces it and allows viewing and concatenation of <code class="language-plaintext highlighter-rouge">gz</code> files.</p> </blockquote><hr /> <h2 id="wordcount-wc--c-m-w-l"> <a href="#wordcount-wc--c-m-w-l" class="anchor-heading" aria-labelledby="wordcount-wc--c-m-w-l"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Wordcount <code class="language-plaintext highlighter-rouge">wc [-c|-m|-w|-l]</code> </h2> <p>The wordcount commands <code class="language-plaintext highlighter-rouge">wc</code> print out the number of bytes <code class="language-plaintext highlighter-rouge">-c</code>, characters <code class="language-plaintext highlighter-rouge">-m</code>, word <code class="language-plaintext highlighter-rouge">-w</code>, and lines <code class="language-plaintext highlighter-rouge">-l</code> within a text file. It is often useful in counting the number of items in your file or a list.</p> <p>One examples is counting the number of genes being analyse in gene count table:</p> <blockquote class="activity"> <p><strong><ins>Try this command:</ins></strong></p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Download Test Data
wget -P ./destination https://raw.githubusercontent.com/gamalielcabria/gamalielcabria.github.io/main/Intro2Bioinfo/files/gene_counts.tsv

# count the number of genes
wc -l gene_counts.tsv
</code></pre></div> </div> </blockquote> <p>Is it the same when you open the file using <code class="language-plaintext highlighter-rouge">less|more</code> and counted the genes?</p> <p class="notice">Files often have a header that you need to account for. It will be counted by <code class="language-plaintext highlighter-rouge">wc</code></p><hr /> <h3 id="grep"> <a href="#grep" class="anchor-heading" aria-labelledby="grep"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> <code class="language-plaintext highlighter-rouge">grep</code> </h3> <p>Search for patterns in a file. Very useful for filtering sequences or logs.</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">grep</span> <span class="s2">"^&gt;"</span> sequences.fasta
<span class="c"># Shows all FASTA headers.</span>
</code></pre></div></div> <h3 id="cut"> <a href="#cut" class="anchor-heading" aria-labelledby="cut"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> <code class="language-plaintext highlighter-rouge">cut</code> </h3> <p>Extract specific columns from tab-delimited files.</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">cut</span> <span class="nt">-f1</span>,3 data.tsv
<span class="c"># Extracts columns 1 and 3 from a tab-separated file.</span>
</code></pre></div></div> <h3 id="paste"> <a href="#paste" class="anchor-heading" aria-labelledby="paste"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> <code class="language-plaintext highlighter-rouge">paste</code> </h3> <p>Combine files line by line.</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">paste </span>genes.txt lengths.txt
<span class="c"># Joins corresponding lines from two files side by side.</span>
</code></pre></div></div> <h3 id="cat"> <a href="#cat" class="anchor-heading" aria-labelledby="cat"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> <code class="language-plaintext highlighter-rouge">cat</code> </h3> <p>Concatenate and display file contents.</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">cat </span>sample1.fq sample2.fq <span class="o">&gt;</span> combined.fq
<span class="c"># Merges two FASTQ files into one.</span>
</code></pre></div></div> <h3 id="sed"> <a href="#sed" class="anchor-heading" aria-labelledby="sed"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> <code class="language-plaintext highlighter-rouge">sed</code> </h3> <p>Stream editor for simple substitutions and line operations.</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">sed</span> <span class="s1">'s/chr/Chr/'</span> genome.txt
<span class="c"># Replaces 'chr' with 'Chr' in each line.</span>
</code></pre></div></div> <h3 id="awk"> <a href="#awk" class="anchor-heading" aria-labelledby="awk"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> <code class="language-plaintext highlighter-rouge">awk</code> </h3> <p>Pattern scanning and processing tool. Useful for complex column-based filtering.</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">awk</span> <span class="s1">'$3 &gt; 100'</span> gene_counts.tsv
<span class="c"># Shows lines where the third column is greater than 100.</span>
</code></pre></div></div> <h3 id="tr"> <a href="#tr" class="anchor-heading" aria-labelledby="tr"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> <code class="language-plaintext highlighter-rouge">tr</code> </h3> <p>Translate or delete characters.</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">tr</span> <span class="s1">'a-z'</span> <span class="s1">'A-Z'</span> &lt; seq.txt
<span class="c"># Converts lowercase bases to uppercase.</span>
</code></pre></div></div> <h1 id="acknowledgement"> <a href="#acknowledgement" class="anchor-heading" aria-labelledby="acknowledgement"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Acknowledgement </h1> <p>This tutorial is adapted from <em>Intro to Unix</em> from Happy Belly Bioinformatics by Michael D. Lee or known as <a href="https://astrobiomike.github.io/unix/">AstroBioMike in github</a>.</p> <h1 id="citation"> <a href="#citation" class="anchor-heading" aria-labelledby="citation"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Citation </h1> <p>[1] Lee, M. (2019). Happy Belly Bioinformatics: an open-source resource dedicated to helping biologists utilize bioinformatics. Journal of Open Source Education, 4(41), 53, https://doi.org/10.21105/jose.00053</p> </main> </div> </div> <div class="search-overlay"></div> </div> </body> </html>
