<!DOCTYPE html> <html lang="en-US"> <head> <meta charset="UTF-8"> <meta http-equiv="X-UA-Compatible" content="IE=Edge"> <link rel="stylesheet" href="/assets/css/just-the-docs-default.css"> <link rel="stylesheet" href="/assets/css/just-the-docs-head-nav.css" id="jtd-head-nav-stylesheet"> <style id="jtd-nav-activation"> .site-nav > ul.nav-list:first-child > li > a, .site-nav > ul.nav-list:first-child > li > ul > li:not(:nth-child(3)) > a, .site-nav > ul.nav-list:first-child > li > ul > li > ul > li a { background-image: none; } .site-nav > ul.nav-list:not(:first-child) a, .site-nav li.external a { background-image: none; } .site-nav > ul.nav-list:first-child > li:nth-child(3) > ul > li:nth-child(3) > a { font-weight: 600; text-decoration: none; }.site-nav > ul.nav-list:first-child > li:nth-child(3) > button svg, .site-nav > ul.nav-list:first-child > li:nth-child(3) > ul > li:nth-child(3) > button svg { transform: rotate(-90deg); }.site-nav > ul.nav-list:first-child > li.nav-list-item:nth-child(3) > ul.nav-list, .site-nav > ul.nav-list:first-child > li.nav-list-item:nth-child(3) > ul.nav-list > li.nav-list-item:nth-child(3) > ul.nav-list { display: block; } </style> <script src="/assets/js/vendor/lunr.min.js"></script> <script src="/assets/js/just-the-docs.js"></script> <meta name="viewport" content="width=device-width, initial-scale=1"> <!-- Begin Jekyll SEO tag v2.8.0 --> <title>Lesson 3 - The commands to rule them all | 🧬 MEB Training Suite</title> <meta name="generator" content="Jekyll v4.3.4" /> <meta property="og:title" content="Lesson 3 - The commands to rule them all" /> <meta property="og:locale" content="en_US" /> <meta name="description" content="Bookmark this to keep an eye on my project updates!" /> <meta property="og:description" content="Bookmark this to keep an eye on my project updates!" /> <link rel="canonical" href="http://localhost:4000/Intro2Bioinfo/Lesson3.html" /> <meta property="og:url" content="http://localhost:4000/Intro2Bioinfo/Lesson3.html" /> <meta property="og:site_name" content="🧬 MEB Training Suite" /> <meta property="og:type" content="website" /> <meta name="twitter:card" content="summary" /> <meta property="twitter:title" content="Lesson 3 - The commands to rule them all" /> <script type="application/ld+json"> {"@context":"https://schema.org","@type":"WebPage","description":"Bookmark this to keep an eye on my project updates!","headline":"Lesson 3 - The commands to rule them all","url":"http://localhost:4000/Intro2Bioinfo/Lesson3.html"}</script> <!-- End Jekyll SEO tag --> </head> <body> <a class="skip-to-main" href="#main-content">Skip to main content</a> <svg xmlns="http://www.w3.org/2000/svg" class="d-none"> <symbol id="svg-link" viewBox="0 0 24 24"> <title>Link</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-link"> <path d="M10 13a5 5 0 0 0 7.54.54l3-3a5 5 0 0 0-7.07-7.07l-1.72 1.71"></path><path d="M14 11a5 5 0 0 0-7.54-.54l-3 3a5 5 0 0 0 7.07 7.07l1.71-1.71"></path> </svg> </symbol> <symbol id="svg-menu" viewBox="0 0 24 24"> <title>Menu</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-menu"> <line x1="3" y1="12" x2="21" y2="12"></line><line x1="3" y1="6" x2="21" y2="6"></line><line x1="3" y1="18" x2="21" y2="18"></line> </svg> </symbol> <symbol id="svg-arrow-right" viewBox="0 0 24 24"> <title>Expand</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-chevron-right"> <polyline points="9 18 15 12 9 6"></polyline> </svg> </symbol> <!-- Feather. MIT License: https://github.com/feathericons/feather/blob/master/LICENSE --> <symbol id="svg-external-link" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-external-link"> <title id="svg-external-link-title">(external link)</title> <path d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"></path><polyline points="15 3 21 3 21 9"></polyline><line x1="10" y1="14" x2="21" y2="3"></line> </symbol> <symbol id="svg-doc" viewBox="0 0 24 24"> <title>Document</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-file"> <path d="M13 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V9z"></path><polyline points="13 2 13 9 20 9"></polyline> </svg> </symbol> <symbol id="svg-search" viewBox="0 0 24 24"> <title>Search</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-search"> <circle cx="11" cy="11" r="8"></circle><line x1="21" y1="21" x2="16.65" y2="16.65"></line> </svg> </symbol> <!-- Bootstrap Icons. MIT License: https://github.com/twbs/icons/blob/main/LICENSE.md --> <symbol id="svg-copy" viewBox="0 0 16 16"> <title>Copy</title> <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-clipboard" viewBox="0 0 16 16"> <path d="M4 1.5H3a2 2 0 0 0-2 2V14a2 2 0 0 0 2 2h10a2 2 0 0 0 2-2V3.5a2 2 0 0 0-2-2h-1v1h1a1 1 0 0 1 1 1V14a1 1 0 0 1-1 1H3a1 1 0 0 1-1-1V3.5a1 1 0 0 1 1-1h1v-1z"/> <path d="M9.5 1a.5.5 0 0 1 .5.5v1a.5.5 0 0 1-.5.5h-3a.5.5 0 0 1-.5-.5v-1a.5.5 0 0 1 .5-.5h3zm-3-1A1.5 1.5 0 0 0 5 1.5v1A1.5 1.5 0 0 0 6.5 4h3A1.5 1.5 0 0 0 11 2.5v-1A1.5 1.5 0 0 0 9.5 0h-3z"/> </svg> </symbol> <symbol id="svg-copied" viewBox="0 0 16 16"> <title>Copied</title> <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-clipboard-check-fill" viewBox="0 0 16 16"> <path d="M6.5 0A1.5 1.5 0 0 0 5 1.5v1A1.5 1.5 0 0 0 6.5 4h3A1.5 1.5 0 0 0 11 2.5v-1A1.5 1.5 0 0 0 9.5 0h-3Zm3 1a.5.5 0 0 1 .5.5v1a.5.5 0 0 1-.5.5h-3a.5.5 0 0 1-.5-.5v-1a.5.5 0 0 1 .5-.5h3Z"/> <path d="M4 1.5H3a2 2 0 0 0-2 2V14a2 2 0 0 0 2 2h10a2 2 0 0 0 2-2V3.5a2 2 0 0 0-2-2h-1v1A2.5 2.5 0 0 1 9.5 5h-3A2.5 2.5 0 0 1 4 2.5v-1Zm6.854 7.354-3 3a.5.5 0 0 1-.708 0l-1.5-1.5a.5.5 0 0 1 .708-.708L7.5 10.793l2.646-2.647a.5.5 0 0 1 .708.708Z"/> </svg> </symbol> </svg> <div class="side-bar"> <div class="site-header" role="banner"> <a href="/" class="site-title lh-tight"> 🧬 MEB Training Suite </a> <button id="menu-button" class="site-button btn-reset" aria-label="Toggle menu" aria-pressed="false"> <svg viewBox="0 0 24 24" class="icon" aria-hidden="true"><use xlink:href="#svg-menu"></use></svg> </button> </div> <nav aria-label="Main" id="site-nav" class="site-nav"> <ul class="nav-list"><li class="nav-list-item"><a href="/" class="nav-list-link">Welcome</a></li><li class="nav-list-item"><a href="/getting-started.html" class="nav-list-link">Getting Started</a></li><li class="nav-list-item"><button class="nav-list-expander btn-reset" aria-label="toggle items in Intro to Bioinfo category" aria-pressed="false"> <svg viewBox="0 0 24 24" aria-hidden="true"><use xlink:href="#svg-arrow-right"></use></svg> </button><a href="/Intro2Bioinfo/" class="nav-list-link">Intro to Bioinfo</a><ul class="nav-list"><li class="nav-list-item"><a href="/Intro2Bioinfo/Lesson1.html" class="nav-list-link">Lesson 1 - The Terminal, the OS, and Bioinfo</a></li><li class="nav-list-item"><a href="/Intro2Bioinfo/Lesson2.html" class="nav-list-link">Lesson 2 - Basic Operations in Terminal and BASH</a></li><li class="nav-list-item"><a href="/Intro2Bioinfo/Lesson3.html" class="nav-list-link">Lesson 3 - The commands to rule them all</a></li><li class="nav-list-item"><a href="/Intro2Bioinfo/Exercise3.html" class="nav-list-link">Exercise 3 - Useful Commands</a></li><li class="nav-list-item"><a href="/Intro2Bioinfo/Lesson4.html" class="nav-list-link">Lesson 4 - Installing New Commands</a></li><li class="nav-list-item"><a href="/Intro2Bioinfo/Lesson5.html" class="nav-list-link">Lesson 5 - BASH Scripting</a></li><li class="nav-list-item"><a href="/Intro2Bioinfo/Lesson6.html" class="nav-list-link">Lesson 6 - Intro to ARC and SLURM JOB submissions</a></li></ul></li></ul> </nav> <footer class="site-footer"> This site uses <a href="https://github.com/just-the-docs/just-the-docs">Just the Docs</a>, a documentation theme for Jekyll. </footer> </div> <div class="main" id="top"> <div id="main-header" class="main-header"> <div class="search" role="search"> <div class="search-input-wrap"> <input type="text" id="search-input" class="search-input" tabindex="0" placeholder="Search 🧬 MEB Training Suite" aria-label="Search 🧬 MEB Training Suite" autocomplete="off"> <label for="search-input" class="search-label"><svg viewBox="0 0 24 24" class="search-icon"><use xlink:href="#svg-search"></use></svg></label> </div> <div id="search-results" class="search-results"></div> </div> <nav aria-label="Auxiliary" class="aux-nav"> <ul class="aux-nav-list"> <li class="aux-nav-list-item"> <a href="https://github.com/gamalielcabria" class="site-button" > Template Repository </a> </li> </ul> </nav> </div> <div class="main-content-wrap"> <nav aria-label="Breadcrumb" class="breadcrumb-nav"> <ol class="breadcrumb-nav-list"> <li class="breadcrumb-nav-list-item"><a href="/Intro2Bioinfo/">Intro to Bioinfo</a></li> <li class="breadcrumb-nav-list-item"><span>Lesson 3 - The commands to rule them all</span></li> </ol> </nav> <div id="main-content" class="main-content"> <main> <h1 id="lesson-3-the-commands-to-rule-them-all"> <a href="#lesson-3-the-commands-to-rule-them-all" class="anchor-heading" aria-labelledby="lesson-3-the-commands-to-rule-them-all"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Lesson 3: The commands to rule them all </h1> <h2 id="basic-bioinformatics-commands"> <a href="#basic-bioinformatics-commands" class="anchor-heading" aria-labelledby="basic-bioinformatics-commands"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Basic bioinformatics commands: </h2> <p>There are tons of built-in commands within Linux OSes that are very useful for bioinformatic analysis. In this section, we will examine a few of the most commonly used commands in bioinformatics. We will adapt content from the <a href="https://astrobiomike.github.io/unix/six-glorious-commands">Happy Belly Bioinformatics</a>.</p> <blockquote class="note"> <p>The following lesson and activity should be tried using a <strong>linux terminal</strong>. Please use your WSL2 and/or register an account to <a href="https://webminal.org/">Webminal</a>.</p> </blockquote><hr /> <h2 id="history-history-n"> <a href="#history-history-n" class="anchor-heading" aria-labelledby="history-history-n"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> History: <code class="language-plaintext highlighter-rouge">history &lt;n&gt;</code> </h2> <p>Look into the recent commands run in the terminal for the past <code class="language-plaintext highlighter-rouge">n</code> lines. Useful for backtracking the commands you have previously run.</p> <blockquote class="activity"> <p><strong><ins>Try this command:</ins></strong></p> <p><code class="language-plaintext highlighter-rouge">history 20</code></p> </blockquote> <p>This should display the last 20 files you have entered in the command-line. If you are searching a specific command, you can also combine <code class="language-plaintext highlighter-rouge">history</code> with <code class="language-plaintext highlighter-rouge">grep</code>(see below) to search a specific command you’ve previously used.</p> <blockquote class="activity"> <p><strong><ins>Try this command:</ins></strong></p> <p><code class="language-plaintext highlighter-rouge">history | grep "ls"</code></p> </blockquote> <p>It should be noted that you can recall also previous commands using your up ⬆️ and down ⬇️ arrow key.</p><hr /> <h2 id="gnu-wget-wget-url"> <a href="#gnu-wget-wget-url" class="anchor-heading" aria-labelledby="gnu-wget-wget-url"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> GNU Wget <code class="language-plaintext highlighter-rouge">wget &lt;url&gt;</code> </h2> <p><code class="language-plaintext highlighter-rouge">wget</code> is a command-line utility for downloading files from the web. It supports HTTP, HTTPS, and FTP protocols and is useful for retrieving datasets, software packages, or scripts directly into your working directory.</p> <blockquote class="activity"> <p><strong><ins>Try this command:</ins></strong></p> <p><code class="language-plaintext highlighter-rouge">wget https://github.com/gamalielcabria/gamalielcabria.github.io/blob/main/Intro2Bioinfo/files/gene_counts.tsv</code></p> </blockquote> <p>The previous command will download the <code class="language-plaintext highlighter-rouge">gene_counts.tsv</code> directly into your current working directory. If you want it to be located in a <strong>different directory</strong>, you can use the command:</p> <p><code class="language-plaintext highlighter-rouge">wget &lt;url&gt; -P &lt;destination path&gt;</code>.</p> <p>To download a folder with multiple files inside, it would be suggested to zip the files or if not possible do a recursive download:</p> <blockquote class="activity"> <p><strong><ins>Try this command:</ins></strong></p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Create a destination folder
mkdir -p ./destination

#download recursively to the destination folder
wget -r -np -nH --cut-dirs=5 -A "*.gtf.gz" ftp://ftp.ensembl.org/pub/release-110/gtf/homo_sapiens/ -P ./destination
</code></pre></div> </div> <blockquote class="info"> <p>This command will download all <code class="language-plaintext highlighter-rouge">*gtf.gz</code> files from ENSEMBL</p> <ul> <li><code class="language-plaintext highlighter-rouge">-r</code>: Recursive download</li> <li><code class="language-plaintext highlighter-rouge">-np</code>: No parent: don’t go to parent directories</li> <li><code class="language-plaintext highlighter-rouge">-nH</code>: No host directory: avoid creating <code class="language-plaintext highlighter-rouge">ftp.ensembl.org/</code></li> <li><code class="language-plaintext highlighter-rouge">--cut-dirs=5</code>: Skip 5 directory levels (so files save cleanly)</li> <li><code class="language-plaintext highlighter-rouge">-A "*.gtf.gz"</code>: Only download files ending in <code class="language-plaintext highlighter-rouge">.gtf.gz</code></li> <li><code class="language-plaintext highlighter-rouge">-P</code>: Destination folder of the file</li> </ul> </blockquote> </blockquote> <p><br /></p><hr /> <h2 id="concatenate-cat"> <a href="#concatenate-cat" class="anchor-heading" aria-labelledby="concatenate-cat"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Concatenate <code class="language-plaintext highlighter-rouge">cat</code> </h2> <p>The <code class="language-plaintext highlighter-rouge">cat</code> (short for <em>concatenate</em>) command is used to display the contents of a file or combine multiple files into one. It’s especially useful for quickly checking small files or merging sequencing data.</p> <div class="language-plaintext activity highlighter-rouge"><div class="highlight"><pre class="highlight"><code>cat file1 file2 &gt; combined_file
# Merges two or more files into one combined_file.
</code></pre></div></div> <p class="note">Without the redirector <code class="language-plaintext highlighter-rouge">&gt;</code> symbol, concatenate <code class="language-plaintext highlighter-rouge">cat</code> will display the concatenated file to <code class="language-plaintext highlighter-rouge">STDOUT</code> only.</p> <blockquote class="activity"> <p><strong><ins>Try this command:</ins></strong></p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Download files to be concatenated
wget -P ./destination https://raw.githubusercontent.com/gamalielcabria/gamalielcabria.github.io/main/Intro2Bioinfo/files/data.tsv
wget -P ./destination https://raw.githubusercontent.com/gamalielcabria/gamalielcabria.github.io/main/Intro2Bioinfo/files/genes.tsv

# Run cat to visually concatenate:
cat data.tsv genes.tsv

# Run it again but different order
cat genes.tsv data.tsv
</code></pre></div> </div> <p>Have you notice any difference?</p> </blockquote><hr /> <h2 id="gunzip-compression-gzip--d-k-file"> <a href="#gunzip-compression-gzip--d-k-file" class="anchor-heading" aria-labelledby="gunzip-compression-gzip--d-k-file"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Gunzip Compression <code class="language-plaintext highlighter-rouge">gzip [-d|-k] &lt;file&gt;</code> </h2> <p><code class="language-plaintext highlighter-rouge">gzip</code> is a widely used command-line tool for compressing text files, such as FASTA, FASTQ, GTF, or VCF files, into <code class="language-plaintext highlighter-rouge">.gz</code> format. This is especially useful in bioinformatics, where raw data files can be very large.</p> <p>This command compresses data.txt into data.txt.gz and removes the original file by default.</p> <blockquote class="activity"> <p><code class="language-plaintext highlighter-rouge">gzip data.txt</code></p> </blockquote> <p>The plaing <code class="language-plaintext highlighter-rouge">gzip</code> command will remove the original file and be replaced with <code class="language-plaintext highlighter-rouge">data.txt.gz</code>. Use <code class="language-plaintext highlighter-rouge">-k</code> to keep the original file:</p> <blockquote class="activity"> <p><code class="language-plaintext highlighter-rouge">gzip -k data.txt</code></p> </blockquote> <p>Meanwhile, decompressing a file uses <code class="language-plaintext highlighter-rouge">-d</code> option or <code class="language-plaintext highlighter-rouge">gunzip</code> command.</p> <blockquote class="activity"> <p><code class="language-plaintext highlighter-rouge">gzip -d data.txt.gz</code></p> </blockquote> <p>To retain the original <code class="language-plaintext highlighter-rouge">gz</code> file:</p> <blockquote class="activity"> <p><code class="language-plaintext highlighter-rouge">gunzip -c data.txt.gz &gt; data.txt</code></p> </blockquote> <p>Compare the sizes of the files between the compressed and uncompressed files. Did you notice any difference?</p> <div class="table-wrapper"><table> <thead> <tr> <th>File Type</th> <th>Compression Ratio</th> <th>Notes</th> </tr> </thead> <tbody> <tr> <td>FASTA</td> <td>2–4× smaller</td> <td>Compresses well due to repetitive sequences</td> </tr> <tr> <td>FASTQ</td> <td>3–6× smaller</td> <td>Especially efficient for raw reads</td> </tr> <tr> <td>GTF/GFF</td> <td>3–5× smaller</td> <td>Mostly text-based, highly compressible</td> </tr> <tr> <td>VCF</td> <td>4–8× smaller</td> <td>Works well but block compression (<code class="language-plaintext highlighter-rouge">bgzip</code>) often preferred</td> </tr> </tbody> </table></div> <blockquote class="notice"> <p>Gunzip compressed files cannot be read by normal text editor or text file view commands such as <code class="language-plaintext highlighter-rouge">more</code>, <code class="language-plaintext highlighter-rouge">cat</code>, and <code class="language-plaintext highlighter-rouge">nano</code>. They will appear gibberish because they are binary. However, <code class="language-plaintext highlighter-rouge">less</code> command are able to interpret them.</p> </blockquote> <blockquote class="note"> <p><strong>zcat</strong> <code class="language-plaintext highlighter-rouge">zcat &lt;gzipped file&gt;</code></p> <p>Given that concatenate <code class="language-plaintext highlighter-rouge">cat</code> command does not work with gzipped files. The <code class="language-plaintext highlighter-rouge">zcat</code> command replaces it and allows viewing and concatenation of <code class="language-plaintext highlighter-rouge">gz</code> files.</p> </blockquote> <p><br /></p><hr /> <h2 id="wordcount-wc--c-m-w-l"> <a href="#wordcount-wc--c-m-w-l" class="anchor-heading" aria-labelledby="wordcount-wc--c-m-w-l"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Wordcount <code class="language-plaintext highlighter-rouge">wc [-c|-m|-w|-l]</code> </h2> <p>The wordcount commands <code class="language-plaintext highlighter-rouge">wc</code> print out the number of bytes <code class="language-plaintext highlighter-rouge">-c</code>, characters <code class="language-plaintext highlighter-rouge">-m</code>, word <code class="language-plaintext highlighter-rouge">-w</code>, and lines <code class="language-plaintext highlighter-rouge">-l</code> within a text file. It is often useful in counting the number of items in your file or a list.</p> <p>One examples is counting the number of genes being analyse in gene count table:</p> <blockquote class="activity"> <p><strong><ins>Try this command:</ins></strong></p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># Download Test Data</span>
wget <span class="nt">-P</span> ./destination https://raw.githubusercontent.com/gamalielcabria/gamalielcabria.github.io/main/Intro2Bioinfo/files/gene_counts.tsv

<span class="c"># count the number of genes</span>
<span class="nb">wc</span> <span class="nt">-l</span> gene_counts.tsv
</code></pre></div> </div> </blockquote> <p>Is it the same when you open the file using <code class="language-plaintext highlighter-rouge">less|more</code> and counted the genes?</p> <p class="notice">Files often have a header that you need to account for. It will be counted by <code class="language-plaintext highlighter-rouge">wc</code></p> <p><br /></p><hr /> <h2 id="column-extraction-command-cut"> <a href="#column-extraction-command-cut" class="anchor-heading" aria-labelledby="column-extraction-command-cut"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Column Extraction Command: <code class="language-plaintext highlighter-rouge">cut</code> </h2> <p>The <code class="language-plaintext highlighter-rouge">cut</code> command is used to extract columns (fields) from tabular files, like <code class="language-plaintext highlighter-rouge">.tsv</code> or <code class="language-plaintext highlighter-rouge">.csv</code>, which is common in bioinformatics data such as gene tables or expression matrices.</p> <blockquote class="activity"> <p><strong><ins>Try this command:</ins></strong></p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>more ./destination/data.tsv
<span class="nb">cut</span> <span class="nt">-f</span> 1 ./destination/data.tsv
<span class="nb">cut</span> <span class="nt">-f</span> 1,3 ./destination/data.tsv
</code></pre></div> </div> </blockquote> <p><code class="language-plaintext highlighter-rouge">cut</code> command cuts the table based on delimiters. The default delimiter is the <strong>tab character</strong> (<code class="language-plaintext highlighter-rouge">\t</code>). To change delimiter, just use the option <code class="language-plaintext highlighter-rouge">-d</code>:</p> <blockquote class="activity"> <p><strong><ins>Try this command:</ins></strong></p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">cut</span> <span class="nt">-d</span><span class="s1">'e'</span> <span class="nt">-f</span> 3 ./destination/data.tsv
</code></pre></div> </div> </blockquote> <p>To permanently combine a cut file, you need to redirect it to an output file.</p> <blockquote class="highlight"> <p><strong><ins>!!DO THIS COMMAND!!</ins></strong></p> <p>Save cut files to different text files</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">cut</span> <span class="nt">-f</span> 1,3 ./destination/data.tsv <span class="o">&gt;</span> score.txt
<span class="nb">cut</span> <span class="nt">-f</span> 1,2 ./destination/data.tsv <span class="o">&gt;</span> value.txt
</code></pre></div> </div> </blockquote> <p><br /></p><hr /> <h2 id="line-by-line-paste-command-paste"> <a href="#line-by-line-paste-command-paste" class="anchor-heading" aria-labelledby="line-by-line-paste-command-paste"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Line-by-line paste command <code class="language-plaintext highlighter-rouge">paste</code> </h2> <p>The <code class="language-plaintext highlighter-rouge">paste</code> command merges lines of multiple files horizontally, inserting a delimiter (default: tab). It’s useful for joining columns from separate files into one table.</p> <blockquote class="activity"> <p><strong><ins>Try this command:</ins></strong></p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Joins corresponding lines from two files side by side.
`paste score.txt value.txt`

# Compare it to the opposite order
`paste value.txt score.txt`
</code></pre></div> </div> </blockquote> <p>Now compare it to a <code class="language-plaintext highlighter-rouge">cat</code> command. What did you observe</p> <p><br /></p><hr /> <h2 id="regular-expression-search-command-grep"> <a href="#regular-expression-search-command-grep" class="anchor-heading" aria-labelledby="regular-expression-search-command-grep"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Regular expression search command <code class="language-plaintext highlighter-rouge">grep</code> </h2> <p><code class="language-plaintext highlighter-rouge">grep</code> is a powerful command-line tool used to search text using <strong>regular expressions (regex)</strong> — patterns that describe sets of strings. It’s widely used in bioinformatics to extract specific sequence IDs, gene names, or annotations from large text-based files.</p> <p>It’s basic command is as such:</p> <blockquote class="activity"> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">grep</span> <span class="s2">"&lt;pattern&gt;"</span> &lt;input file&gt;
</code></pre></div> </div> </blockquote> <p><br /> What is regular expression?</p> <blockquote class="highlight"> <h3 id="-regular-expressions-regex"> <a href="#-regular-expressions-regex" class="anchor-heading" aria-labelledby="-regular-expressions-regex"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> 🧩 Regular Expressions (Regex) </h3> <p>Regular expressions are powerful patterns used to match strings of text. They are essential for searching, filtering, and extracting data in large bioinformatics files like FASTA, GTF, or VCF.</p> <p>For example:</p> <ul> <li><code class="language-plaintext highlighter-rouge">^gene</code> — matches lines that <strong>start(^)</strong> with “gene”</li> <li><code class="language-plaintext highlighter-rouge">\.gz$</code> — matches filenames that <strong>end($)</strong> with <code class="language-plaintext highlighter-rouge">.gz</code>. <blockquote> <ul> <li>The backward slash <code class="language-plaintext highlighter-rouge">\</code> here forces the program to treat <code class="language-plaintext highlighter-rouge">.</code> as a <strong>dot</strong> character</li> <li>Otherwise, <code class="language-plaintext highlighter-rouge">.</code> in regex can signal any character or whitespace</li> </ul> </blockquote> </li> <li><code class="language-plaintext highlighter-rouge">[ATCG]{10,}</code> — matches sequences with <strong>10 or more</strong> A/T/C/G characters</li> </ul> <p>Regex is used with tools like <code class="language-plaintext highlighter-rouge">grep</code>, <code class="language-plaintext highlighter-rouge">sed</code>, <code class="language-plaintext highlighter-rouge">awk</code>, and even within programming languages like Python or R.</p> <blockquote class="note"> <p>Learn more about how regular expressions work using the following <a href="https://media.datacamp.com/legacy/image/upload/v1665049611/Marketing/Blog/Regular_Expressions_Cheat_Sheet.pdf">cheatsheet</a>.</p> <p>If you want to check out and practice how it works, visit the website <a href="https://regex101.com/">regex101.com</a>.</p> <p>For now let us stick to the basics.</p> </blockquote> </blockquote> <p><br /> <br /> <strong>GREP</strong> most known function in bioinformatics is the counting of reads in a fasta file. If you recall, fasta file consists of a header starting with <code class="language-plaintext highlighter-rouge">&gt;</code> followed by some text and a sequence file. To know the number of fasta sequences within a fasta file, simply use the following:</p> <blockquote class="activity"> <p><strong><ins>Try this command:</ins></strong></p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c">#Download a fasta file</span>
wget <span class="nt">-P</span> ./destination https://raw.githubusercontent.com/gamalielcabria/gamalielcabria.github.io/main/Intro2Bioinfo/files/random.fasta

<span class="c">#Count the number of headers using `-c` option</span>
<span class="nb">grep</span> <span class="nt">-c</span> <span class="s2">"^&gt;"</span> ./destination/random.fasta
</code></pre></div> </div> </blockquote> <p>The <code class="language-plaintext highlighter-rouge">-c</code> is used to count the number of reoccurrence of the pattern within <code class="language-plaintext highlighter-rouge">"&lt;pattern&gt;"</code>. Again, <code class="language-plaintext highlighter-rouge">^</code> signifies start of a line. Thus, <code class="language-plaintext highlighter-rouge">"^&gt;"</code> counts the number of <code class="language-plaintext highlighter-rouge">&gt;</code> in the file that is at the start of the line (i.e. headers).</p> <p>Can you compare it without the <code class="language-plaintext highlighter-rouge">-c</code> option?</p> <blockquote class="activity"> <p><strong><ins>Try this command:</ins></strong></p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">grep</span> <span class="s2">"^&gt;"</span> ./destination/random.fasta
</code></pre></div> </div> </blockquote> <p>It should display the lines that match and color the pattern that matches. <br /> <br /> <br /> <strong>GREP</strong> can be used only on text files. To read compressed or binary files such as <code class="language-plaintext highlighter-rouge">*.gz</code>. However, you do not to decompress to process them but you can use pipe to chain the commands. An example for counting all non-comment files in <code class="language-plaintext highlighter-rouge">gtf</code> file:</p> <blockquote class="activity"> <p><strong><ins>Try this command:</ins></strong></p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>zcat ./destination/Homo_sapiens.GRCh38.110.gtf.gz | <span class="nb">grep</span> <span class="nt">-c</span> <span class="nt">-v</span> <span class="s2">"^#"</span>
</code></pre></div> </div> </blockquote> <p>The <code class="language-plaintext highlighter-rouge">gtf</code> file often has commented text in top of the file such as this:</p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>#!genome-build GRCh38.p14
#!genome-version GRCh38
#!genome-date 2013-12
#!genome-build-accession GCA_000001405.29
#!genebuild-last-updated 2023-03
1       havana  gene    182696  184174  .       +       .       gene_id "ENSG00000279928"; gene_version "2"; gene_name "DDX11L17"; gene_source "havana"; gene_biotype "unprocessed_pseudogene";
1       havana  transcript  182696      184174  .       +       .       gene_id "ENSG00000279928"; gene_version "2"; transcript_id "ENST00000624431"; transcript_version "2"; gene_name "DDX11L17"; gene_source "havana"; gene_biotype "unprocessed_pseudogene"; transcript_name "DDX11L17-201"; transcript_source "havana"; transcript_biotype "unprocessed_pseudogene"; tag "basic"; tag "Ensembl_canonical"; transcript_support_level "NA";
1       havana  exon    182696  182746  .       +       .       gene_id "ENSG00000279928"; gene_version "2"; transcript_id "ENST00000624431"; transcript_version "2"; exon_number "1"; gene_name "DDX11L17"; gene_source "havana"; gene_biotype "unprocessed_pseudogene"; transcript_name "DDX11L17-201"; transcript_source "havana"; transcript_biotype "unprocessed_pseudogene"; exon_id "ENSE00003759020"; exon_version "2"; tag "basic"; tag "Ensembl_canonical"; transcript_support_level "NA";
</code></pre></div></div> <p>The <code class="language-plaintext highlighter-rouge">-v</code> option in the <code class="language-plaintext highlighter-rouge">grep</code> command skip all files that matches the pattern: lines that starts with <code class="language-plaintext highlighter-rouge">#</code>.</p> <p>We can also chain <code class="language-plaintext highlighter-rouge">grep</code> commands to search more thoroughly within the file. For example, if we want to find a line that is a <strong>rRNA gene</strong> and its <strong>gene</strong> description in the <code class="language-plaintext highlighter-rouge">gtf</code> file, we can run this:</p> <div class="language-bash activity highlighter-rouge"><div class="highlight"><pre class="highlight"><code>zcat ./destination/Homo_sapiens.GRCh38.110.gtf.gz | <span class="nb">grep</span> <span class="nt">-P</span> <span class="s1">'\tgene\t'</span> file.gtf | <span class="nb">grep</span> <span class="s1">'gene_biotype "rRNA"'</span>
</code></pre></div></div> <p>Now, can you count the number of <strong>tRNA exons</strong> in these gtf file?</p><hr /> <h2 id="the-stream-editor-command-sed"> <a href="#the-stream-editor-command-sed" class="anchor-heading" aria-labelledby="the-stream-editor-command-sed"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> The Stream Editor Command <code class="language-plaintext highlighter-rouge">sed</code> </h2> <p>The <code class="language-plaintext highlighter-rouge">sed</code> command is a powerful tool for editing text in files or streams. It’s often used for find-and-replace operations, especially when working with large datasets or automated scripts.</p> <blockquote class="activity"> <p><code class="language-plaintext highlighter-rouge">sed [options: -i|-E] 's/&lt;old pattern or text&gt;/&lt;new replacement&gt;/g' &lt;file&gt;</code></p> </blockquote> <p>Several options can be used when running <code class="language-plaintext highlighter-rouge">sed</code>, check <code class="language-plaintext highlighter-rouge">--help</code> for more info. Without any options declared, the <code class="language-plaintext highlighter-rouge">sed</code> output will just be displayed in <code class="language-plaintext highlighter-rouge">STDOUT</code>. The command <code class="language-plaintext highlighter-rouge">-i</code>, means in-line replacement, can be use to edit the input file without displaying.</p> <p class="info"><code class="language-plaintext highlighter-rouge">sed</code> command utilizes <em>regular expression</em> for matching text. The <code class="language-plaintext highlighter-rouge">sed</code> has so many options than what I can easily discuss in this beginner guide. Please look up this <strong><a href="https://quickref.me/sed.html">SED cheatsheet</a></strong> for more details.</p> <p>One example of <code class="language-plaintext highlighter-rouge">sed</code> usage in bioinformatics is quick renameing of headers of fasta file.</p> <blockquote class="activity"> <p><strong><ins>Try this command:</ins></strong></p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># Visualize the content of fasta file</span>
<span class="nb">head</span> ./destination/random.fasta

<span class="c"># Let us replace the header name `seq` by including the species they are from</span>
<span class="nb">sed</span> <span class="s1">'s/^&gt;seq/&gt;Bacillus subtillis \| gene /g'</span> ./destination/random.fasta
</code></pre></div> </div> </blockquote> <p>Did it change the contents of the file? What do you need to do to change the content of the file <strong>permanently</strong>?</p> <p><br /> <code class="language-plaintext highlighter-rouge">sed</code> is also useful when removing or replacing unwanted targets. For example, I want to remove sequences that has adenosine homopolymer count of <strong>4 or more</strong> and replace them with <strong>N’s</strong>:</p> <blockquote class="activity"> <p><strong><ins>Try this command:</ins></strong></p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># See how many matches using `grep`</span>
<span class="nb">grep</span> <span class="nt">-E</span> <span class="s2">"A{4,}"</span> ./destination/rando.fasta
<span class="c"># run replacement and visualize it using</span>
<span class="nb">sed</span> <span class="nt">-E</span> <span class="s1">'s/A{4,}/NNNNNNNN/g'</span> ./destination/random.fasta | <span class="nb">grep</span> <span class="nt">-E</span> <span class="s2">"N{8}"</span> <span class="nt">-B1</span> 
</code></pre></div> </div> <blockquote class="notice"> <p>Remember this does not invoke <code class="language-plaintext highlighter-rouge">-i</code> which means the sed replacement is not save in the file.</p> </blockquote> </blockquote> <p>Lastly, one other use of <code class="language-plaintext highlighter-rouge">sed</code> is changing file delimtiers. Converting a <code class="language-plaintext highlighter-rouge">tsv</code> file to <code class="language-plaintext highlighter-rouge">csv</code> in which the <code class="language-plaintext highlighter-rouge">\t</code> is replaced by <code class="language-plaintext highlighter-rouge">,</code>:</p> <blockquote class="highlight"> <p><strong><ins>!!DO THIS COMMAND!!</ins></strong></p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">sed</span> <span class="s1">'s/\t/,/g'</span> data.tsv <span class="o">&gt;</span> data.csv
</code></pre></div> </div> </blockquote> <p><br /></p><hr /> <h2 id="pattern-scanning-and-text-processing-language-awk"> <a href="#pattern-scanning-and-text-processing-language-awk" class="anchor-heading" aria-labelledby="pattern-scanning-and-text-processing-language-awk"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Pattern scanning and text processing language <code class="language-plaintext highlighter-rouge">awk</code> </h2> <p><code class="language-plaintext highlighter-rouge">awk</code> is a powerful command-line tool for pattern scanning, text processing, and column-wise operations in structured files like TSV or CSV.</p> <p>It reads input line by line, splits lines into fields, and lets you perform calculations, filtering, or formatting.</p> <blockquote class="activity"> <p><code class="language-plaintext highlighter-rouge">awk '$3 &gt; 150' gene_counts.tsv</code></p> <blockquote> <ul> <li>This prints lines from gene_counts.tsv where the third column is greater than 100.</li> </ul> </blockquote> </blockquote> <p>It can also perform more complicated task such as filtering based on values for multiple columns:</p> <blockquote class="activity"> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">awk</span> <span class="s1">'
BEGIN {OFS = "\t"}
$2 &gt; 180 &amp;&amp; $7 &lt; 300 { print }
'</span> &lt; ./destination/gene_counts.tsv
</code></pre></div> </div> </blockquote> <p><br /> <code class="language-plaintext highlighter-rouge">awk</code> is indispensable in bioinformatics pipelines for filtering large tabular files like GTFs, count matrices, and sequence stats.</p> <h3 id="other-common-uses"> <a href="#other-common-uses" class="anchor-heading" aria-labelledby="other-common-uses"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Other common uses: </h3> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>awk '{print $1, $3}' data.tsv         # Print columns 1 and 3
awk -F',' '{print $2}' data.csv       # Use comma as delimiter
awk 'NR &gt; 1' file.tsv                 # Skip header (print after 1st row)
</code></pre></div></div> <p>Remember earlier that we count the rows of the table using <code class="language-plaintext highlighter-rouge">wc</code> command? Now, we can also combine it with awk to count the number of columns of <code class="language-plaintext highlighter-rouge">data.csv</code></p> <blockquote class="activity"> <p><strong><ins>Try this command:</ins></strong></p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>wc -l &lt; gene_counts.tsv
head -1 gene_counts.tsv | awk -F'\t' '{print NF}'
</code></pre></div> </div> <ul> <li><code class="language-plaintext highlighter-rouge">wc -l</code>: counts the number of lines (i.e., rows)</li> <li><code class="language-plaintext highlighter-rouge">head -1</code>: grabs the header row</li> <li><code class="language-plaintext highlighter-rouge">awk -F'\t' '{print NF}'</code>: prints the number of fields/<strong>NF</strong> (columns), assuming tab-delimited</li> </ul> </blockquote> <p><br /></p><hr /> <h2 id="translate-tr"> <a href="#translate-tr" class="anchor-heading" aria-labelledby="translate-tr"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Translate <code class="language-plaintext highlighter-rouge">tr</code> </h2> <p>The <code class="language-plaintext highlighter-rouge">tr</code> (translate) command is used to replace, squeeze, or delete characters from input. It works only on character-by-character transformations and reads from standard input.</p> <p>An example, let us change the <strong>[AGCT]</strong> nucleotides in this fasta to lowercase. We can use:</p> <blockquote class="activity"> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">tr</span> <span class="s1">'A-Z'</span> <span class="s1">'a-z'</span> &lt; random.fasta
</code></pre></div> </div> </blockquote> <p>We can use <code class="language-plaintext highlighter-rouge">tr</code> also to remove all digits in the file</p> <blockquote class="activity"> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">tr</span> <span class="nt">-d</span> <span class="s1">'0-9'</span> &lt; random.fasta
</code></pre></div> </div> </blockquote> <p><br /></p><hr /> <h2 id="display-text-or-variable-command-echo"> <a href="#display-text-or-variable-command-echo" class="anchor-heading" aria-labelledby="display-text-or-variable-command-echo"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Display text or variable command <code class="language-plaintext highlighter-rouge">echo</code> </h2> <p>The <code class="language-plaintext highlighter-rouge">echo</code> command prints text or variables to the terminal. It’s often used in scripts to show messages, output variable values, or create files on the fly.</p> <ul> <li><strong>Prints a simple message</strong>:</li> </ul> <blockquote class="activity"> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">echo</span> <span class="s2">"Hello, bioinformaticians!"</span>
</code></pre></div> </div> </blockquote> <ul> <li><strong>Show the value of a variable</strong>:</li> </ul> <blockquote class="activity"> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">sample</span><span class="o">=</span><span class="s2">"sample1"</span>
<span class="nb">echo</span> <span class="s2">"Currently processing: </span><span class="nv">$sample</span><span class="s2">"</span>
</code></pre></div> </div> <p>Values to the <strong>variables</strong> in bash is assigned using the <code class="language-plaintext highlighter-rouge">=</code> symbol. The <strong>variable</strong> name to the left of it and the values from the right.</p> </blockquote> <p class="note">There should be no space between the <strong>variable</strong>, <code class="language-plaintext highlighter-rouge">=</code>, and the assigned <strong>value</strong> to the left.</p> <p>We can use it also to write FASTA or any text to a file!</p> <ul> <li><strong>Write FASTA content to a file</strong>:</li> </ul> <blockquote class="activity"> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">echo</span> <span class="nt">-e</span> <span class="s2">"&gt;geneX</span><span class="se">\n</span><span class="s2">ATGCGTACGTA"</span> <span class="o">&gt;</span> geneX.fasta
</code></pre></div> </div> </blockquote> <ul> <li><strong>Add content to a file</strong>:</li> </ul> <blockquote class="activity"> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">echo</span> <span class="nt">-e</span> <span class="s2">"&gt;geneX</span><span class="se">\n</span><span class="s2">ATGCGTACGTA"</span> <span class="o">&gt;</span> geneX.fasta
</code></pre></div> </div> </blockquote> <p><br /> <br /></p> <h1 id="summary"> <a href="#summary" class="anchor-heading" aria-labelledby="summary"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Summary </h1> <div class="table-wrapper"><table> <thead> <tr> <th>Command</th> <th>Description</th> <th>Example</th> <th>Use Case</th> </tr> </thead> <tbody> <tr> <td><code class="language-plaintext highlighter-rouge">history</code></td> <td>Shows recently used commands</td> <td><code class="language-plaintext highlighter-rouge">history 20</code></td> <td>Recall past terminal commands</td> </tr> <tr> <td><code class="language-plaintext highlighter-rouge">wget</code></td> <td>Downloads files from the web</td> <td><code class="language-plaintext highlighter-rouge">wget &lt;url&gt;</code></td> <td>Retrieve datasets/scripts</td> </tr> <tr> <td><code class="language-plaintext highlighter-rouge">cat</code></td> <td>Displays or combines files</td> <td><code class="language-plaintext highlighter-rouge">cat file1 file2 &gt; merged.txt</code></td> <td>Merge sequencing files</td> </tr> <tr> <td><code class="language-plaintext highlighter-rouge">gzip</code> / <code class="language-plaintext highlighter-rouge">gunzip</code></td> <td>Compress/decompress files</td> <td><code class="language-plaintext highlighter-rouge">gzip file.txt</code> / <code class="language-plaintext highlighter-rouge">gunzip file.gz</code></td> <td>Save storage on large files</td> </tr> <tr> <td><code class="language-plaintext highlighter-rouge">wc</code></td> <td>Counts lines, words, etc.</td> <td><code class="language-plaintext highlighter-rouge">wc -l gene_counts.tsv</code></td> <td>Count rows in tables</td> </tr> <tr> <td><code class="language-plaintext highlighter-rouge">cut</code></td> <td>Extracts columns</td> <td><code class="language-plaintext highlighter-rouge">cut -f1,3 data.tsv</code></td> <td>Subset tabular data</td> </tr> <tr> <td><code class="language-plaintext highlighter-rouge">paste</code></td> <td>Merges files line-by-line</td> <td><code class="language-plaintext highlighter-rouge">paste file1 file2</code></td> <td>Combine columns from files</td> </tr> <tr> <td><code class="language-plaintext highlighter-rouge">grep</code></td> <td>Searches using regex</td> <td><code class="language-plaintext highlighter-rouge">grep "^&gt;" fasta.fna</code></td> <td>Count sequences or search patterns</td> </tr> <tr> <td><code class="language-plaintext highlighter-rouge">sed</code></td> <td>Stream editor for text</td> <td><code class="language-plaintext highlighter-rouge">sed 's/old/new/g' file.txt</code></td> <td>Replace or clean text</td> </tr> <tr> <td><code class="language-plaintext highlighter-rouge">awk</code></td> <td>Processes text by column</td> <td><code class="language-plaintext highlighter-rouge">awk '$2 &gt; 100' file.tsv</code></td> <td>Filter tabular data</td> </tr> <tr> <td><code class="language-plaintext highlighter-rouge">tr</code></td> <td>Translates/deletes characters</td> <td><code class="language-plaintext highlighter-rouge">tr 'A-Z' 'a-z' &lt; file</code></td> <td>Change case or strip characters</td> </tr> <tr> <td><code class="language-plaintext highlighter-rouge">echo</code></td> <td>Prints text/variables</td> <td><code class="language-plaintext highlighter-rouge">echo "Hello!"</code></td> <td>Display info or write files</td> </tr> </tbody> </table></div> <h1 id="acknowledgement"> <a href="#acknowledgement" class="anchor-heading" aria-labelledby="acknowledgement"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Acknowledgement </h1> <p>This tutorial is adapted from <em>Intro to Unix</em> from Happy Belly Bioinformatics by Michael D. Lee or known as <a href="https://astrobiomike.github.io/unix/">AstroBioMike in github</a>.</p> <h1 id="citation"> <a href="#citation" class="anchor-heading" aria-labelledby="citation"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Citation </h1> <p>[1] Lee, M. (2019). Happy Belly Bioinformatics: an open-source resource dedicated to helping biologists utilize bioinformatics. Journal of Open Source Education, 4(41), 53, https://doi.org/10.21105/jose.00053</p> </main> </div> </div> <div class="search-overlay"></div> </div> </body> </html>
